// mson-react/lib/fields/autocomplete-select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _reactSelect = _interopRequireDefault(require("react-select"));

var _styles = require("@material-ui/core/styles");

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _TextField = _interopRequireDefault(require("@material-ui/core/TextField"));

var _Paper = _interopRequireDefault(require("@material-ui/core/Paper"));

var _Chip = _interopRequireDefault(require("@material-ui/core/Chip"));

var _MenuItem = _interopRequireDefault(require("@material-ui/core/MenuItem"));

var _icon = _interopRequireDefault(require("../icon"));

var _colorManipulator = require("@material-ui/core/styles/colorManipulator");

var styles = function styles(theme) {
  return {
    input: {
      display: 'flex',
      padding: 0,
      minWidth: 240 // A good default for most labels

    },
    valueContainer: {
      display: 'flex',
      flexWrap: 'wrap',
      flex: 1,
      alignItems: 'center',
      // Adjust for MUI label
      marginTop: theme.spacing.unit * 2 + 5
    },
    chip: {
      margin: "".concat(theme.spacing.unit / 1, "px ").concat(theme.spacing.unit / 4, "px")
    },
    chipFocused: {
      backgroundColor: (0, _colorManipulator.emphasize)(theme.palette.type === 'light' ? theme.palette.grey[300] : theme.palette.grey[700], 0.08)
    },
    noOptionsMessage: {
      padding: "".concat(theme.spacing.unit, "px ").concat(theme.spacing.unit * 2, "px")
    },
    singleValue: {
      fontSize: theme.typography.subtitle1.fontSize
    },
    // placeholder: {
    //   position: 'absolute',
    //   left: 2,
    //   fontSize: theme.typography.subtitle1.fontSize
    // },
    paper: {
      position: 'absolute',
      zIndex: 1,
      // Adjust for MUI label
      // marginTop: theme.spacing.unit * 7
      marginTop: theme.spacing.unit * 1
    },
    root: {
      // Allow for help icon to be placed on right of field
      display: 'inline-flex'
    },
    fullWidth: {
      width: '100%'
    },
    disabled: {
      color: theme.palette.text.secondary
    }
  };
};

function NoOptionsMessage(props) {
  return _react.default.createElement(_Typography.default, (0, _extends2.default)({
    color: "textSecondary",
    className: props.selectProps.classes.noOptionsMessage
  }, props.innerProps), props.children);
}

function inputComponent(_ref) {
  var inputRef = _ref.inputRef,
      props = (0, _objectWithoutProperties2.default)(_ref, ["inputRef"]);
  return _react.default.createElement("div", (0, _extends2.default)({
    ref: inputRef
  }, props));
}

function Control(props) {
  var optionalProps = {};

  if (props.selectProps.fullWidth) {
    optionalProps.fullWidth = true;
  }

  return _react.default.createElement(_TextField.default, (0, _extends2.default)({}, optionalProps, {
    InputProps: {
      inputComponent: inputComponent,
      inputProps: (0, _objectSpread2.default)({
        className: props.selectProps.classes.input,
        inputRef: props.innerRef,
        children: props.children
      }, props.innerProps)
    }
  }, props.selectProps.textFieldProps));
}

function Option(props) {
  return _react.default.createElement(_MenuItem.default, (0, _extends2.default)({
    buttonRef: props.innerRef,
    selected: props.isFocused,
    component: "div",
    style: {
      fontWeight: props.isSelected ? 500 : 400
    }
  }, props.innerProps), props.children);
} // Note: kept in case we want to implement the placeholder in the future
// function Placeholder(props) {
//   return (
//     <Typography
//       color="textSecondary"
//       className={props.selectProps.classes.placeholder}
//       {...props.innerProps}
//     >
//       {props.children}
//     </Typography>
//   );
// }


function SingleValue(props) {
  return _react.default.createElement(_Typography.default, (0, _extends2.default)({
    className: (0, _classnames.default)(props.selectProps.classes.singleValue, props.selectProps.isDisabled ? props.selectProps.classes.disabled : null)
  }, props.innerProps), props.children);
}

function ValueContainer(props) {
  return _react.default.createElement("div", {
    className: props.selectProps.classes.valueContainer
  }, props.children);
}

function MultiValue(props) {
  var _classNames;

  return _react.default.createElement(_Chip.default, {
    tabIndex: -1,
    label: props.children,
    className: (0, _classnames.default)(props.selectProps.classes.chip, (_classNames = {}, (0, _defineProperty2.default)(_classNames, props.selectProps.classes.chipFocused, props.isFocused), (0, _defineProperty2.default)(_classNames, props.selectProps.classes.disabled, props.selectProps.isDisabled), _classNames)),
    onDelete: props.removeProps.onClick,
    deleteIcon: _react.default.createElement(_icon.default, {
      icon: "Cancel"
    })
  });
}

function Menu(props) {
  return _react.default.createElement(_Paper.default, (0, _extends2.default)({
    square: true,
    className: props.selectProps.classes.paper
  }, props.innerProps), props.children);
}

var components = {
  Control: Control,
  Menu: Menu,
  MultiValue: MultiValue,
  NoOptionsMessage: NoOptionsMessage,
  Option: Option,
  // Placeholder,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer
};

var AutoCompleteSelect =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(AutoCompleteSelect, _React$Component);

  function AutoCompleteSelect() {
    (0, _classCallCheck2.default)(this, AutoCompleteSelect);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(AutoCompleteSelect).apply(this, arguments));
  }

  (0, _createClass2.default)(AutoCompleteSelect, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          classes = _this$props.classes,
          options = _this$props.options,
          isClearable = _this$props.isClearable,
          placeholder = _this$props.placeholder,
          onChange = _this$props.onChange,
          onBlur = _this$props.onBlur,
          onFocus = _this$props.onFocus,
          onInputChange = _this$props.onInputChange,
          value = _this$props.value,
          isDisabled = _this$props.isDisabled,
          fullWidth = _this$props.fullWidth,
          isMulti = _this$props.isMulti;
      var selectStyles = {
        dropdownIndicator: function dropdownIndicator(base) {
          return (0, _objectSpread2.default)({}, base, {
            cursor: 'pointer'
          });
        },
        clearIndicator: function clearIndicator(base) {
          return (0, _objectSpread2.default)({}, base, {
            cursor: 'pointer'
          });
        },
        // Needed so that the menu appears above any dialog windows
        menuPortal: function menuPortal(base) {
          return (0, _objectSpread2.default)({}, base, {
            zIndex: 2000
          });
        }
      };
      return _react.default.createElement(_reactSelect.default, {
        className: (0, _classnames.default)(classes.root, fullWidth && classes.fullWidth),
        classes: classes,
        styles: selectStyles,
        options: options,
        components: components,
        value: value,
        onChange: onChange,
        onBlur: onBlur,
        onFocus: onFocus,
        onInputChange: onInputChange,
        placeholder: placeholder,
        isClearable: isClearable,
        isDisabled: isDisabled,
        fullWidth: fullWidth,
        isMulti: isMulti // Needed so that the menu is not clipped by the app's root level overflow:hidden or when it
        // appears in a dialog window
        ,
        menuPortalTarget: document.body
      });
    }
  }]);
  return AutoCompleteSelect;
}(_react.default.Component);

var _default = (0, _styles.withStyles)(styles)(AutoCompleteSelect);

exports.default = _default;
