// mson-react/lib/fields/boolean-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _Switch = _interopRequireDefault(require("@material-ui/core/Switch"));

var _commonField = _interopRequireDefault(require("./common-field"));

var _attach = _interopRequireDefault(require("../attach"));

var _FormControlLabel = _interopRequireDefault(require("@material-ui/core/FormControlLabel"));

var _displayValueTypography = _interopRequireDefault(require("./display-value-typography"));

var BooleanField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(BooleanField, _React$PureComponent);

  function BooleanField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, BooleanField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(BooleanField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (event) {
      _this.props.component.setValue(event.target.checked);
    });
    return _this;
  }

  (0, _createClass2.default)(BooleanField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          value = _this$props.value,
          disabled = _this$props.disabled,
          component = _this$props.component,
          editable = _this$props.editable,
          useDisplayValue = _this$props.useDisplayValue,
          label = _this$props.label;
      var hideLabelUI = null;
      var fld = null;

      if (editable && !useDisplayValue) {
        hideLabelUI = true;
        fld = _react.default.createElement(_FormControlLabel.default, {
          control: _react.default.createElement(_Switch.default, {
            checked: value ? value : false,
            onChange: this.handleChange,
            value: "true",
            disabled: disabled
          }),
          label: label
        });
      } else {
        fld = _react.default.createElement(_displayValueTypography.default, null, component.getDisplayValue());
      }

      return _react.default.createElement(_commonField.default, {
        component: component,
        hideLabelUI: hideLabelUI
      }, fld);
    }
  }]);
  return BooleanField;
}(_react.default.PureComponent);

var _default = (0, _attach.default)(['value', 'err', 'disabled', 'editable', 'useDisplayValue', 'label'])(BooleanField);

exports.default = _default;
