// mson-react/lib/fields/button-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("../attach"));

var _button = _interopRequireDefault(require("../button"));

var ButtonField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(ButtonField, _React$PureComponent);

  function ButtonField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ButtonField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ButtonField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function () {
      var type = _this.props.type; // Is the button not a submit button? Let the form handle submit buttons so that the form
      // performs the validation

      if (type !== 'submit') {
        _this.props.component.emitClick();
      }
    });
    return _this;
  }

  (0, _createClass2.default)(ButtonField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          label = _this$props.label,
          type = _this$props.type,
          disabled = _this$props.disabled,
          icon = _this$props.icon,
          fullWidth = _this$props.fullWidth,
          variant = _this$props.variant;
      return _react.default.createElement(_button.default, {
        type: type,
        label: label,
        disabled: disabled,
        onClick: this.handleClick,
        icon: icon,
        fullWidth: fullWidth,
        variant: variant
      });
    }
  }]);
  return ButtonField;
}(_react.default.PureComponent);

var _default = (0, _attach.default)(['label', 'type', 'disabled', 'icon', 'fullWidth', 'variant'])(ButtonField);

exports.default = _default;
