// mson-react/lib/fields/component-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("../attach"));

var _component = _interopRequireDefault(require("../component"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = function styles(theme) {
  return {
    root: {
      marginLeft: theme.spacing.unit
    }
  };
};

var ComponentField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(ComponentField, _React$PureComponent);

  function ComponentField() {
    (0, _classCallCheck2.default)(this, ComponentField);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ComponentField).apply(this, arguments));
  }

  (0, _createClass2.default)(ComponentField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          content = _this$props.content,
          classes = _this$props.classes;
      return _react.default.createElement("div", {
        className: classes.root
      }, _react.default.createElement(_component.default, {
        component: content
      }));
    }
  }]);
  return ComponentField;
}(_react.default.PureComponent);

ComponentField = (0, _withStyles.default)(styles)(ComponentField);
ComponentField = (0, _attach.default)(['content'])(ComponentField);
var _default = ComponentField;
exports.default = _default;
