// mson-react/lib/fields/date-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("../attach"));

var _commonField = _interopRequireDefault(require("./common-field"));

var _MuiPickersUtilsProvider = _interopRequireDefault(require("material-ui-pickers/MuiPickersUtilsProvider"));

var _dateFns = _interopRequireDefault(require("@date-io/date-fns"));

var _DateTimePicker = _interopRequireDefault(require("material-ui-pickers/DateTimePicker"));

var _DatePicker = _interopRequireDefault(require("material-ui-pickers/DatePicker"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _displayValueTypography = _interopRequireDefault(require("./display-value-typography"));

var styles = function styles(theme) {
  return {
    root: {
      // We need to add a margin as the picker doesn't play well with the Material-UI label
      marginTop: theme.spacing.unit * 2
    }
  };
};

var DateField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(DateField, _React$PureComponent);

  function DateField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, DateField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(DateField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDateChange", function (date) {
      _this.props.component.setValue(date);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function () {
      _this.props.component.setTouched(true);
    });
    return _this;
  }

  (0, _createClass2.default)(DateField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          component = _this$props.component,
          classes = _this$props.classes,
          value = _this$props.value,
          includeTime = _this$props.includeTime,
          editable = _this$props.editable,
          useDisplayValue = _this$props.useDisplayValue,
          minDate = _this$props.minDate,
          maxDate = _this$props.maxDate,
          fullWidth = _this$props.fullWidth,
          disabled = _this$props.disabled,
          accessEditable = _this$props.accessEditable;
      var dis = accessEditable === false || disabled;
      var shrinkLabel = false;
      var Component = includeTime ? _DateTimePicker.default : _DatePicker.default;
      var fld = null;

      if (editable && !useDisplayValue) {
        // The picker doesn't play well with the label from Material-UI so we need to manually shrink
        // the label when there is a value.
        shrinkLabel = !!value;
        fld = _react.default.createElement(_MuiPickersUtilsProvider.default, {
          utils: _dateFns.default
        }, _react.default.createElement("span", null, _react.default.createElement(Component // Value cannot be undefined or else picker will default to today
        , {
          value: value ? value : null,
          onChange: this.handleDateChange,
          clearable: true,
          className: classes.root,
          minDate: minDate,
          maxDate: maxDate,
          onClose: this.handleBlur,
          fullWidth: fullWidth,
          disabled: dis // format="M/d/YYYY h:m a"

        })));
      } else {
        fld = _react.default.createElement(_displayValueTypography.default, null, component.getDisplayValue());
      }

      return _react.default.createElement(_commonField.default, {
        component: component,
        shrinkLabel: shrinkLabel
      }, fld);
    }
  }]);
  return DateField;
}(_react.default.PureComponent);

DateField = (0, _withStyles.default)(styles)(DateField);

var _default = (0, _attach.default)(['value', 'includeTime', 'editable', 'useDisplayValue', 'minDate', 'maxDate', 'fullWidth', 'disabled'])(DateField);

exports.default = _default;
