// mson-react/lib/fields/display-value-typography.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = function styles(theme) {
  return {
    root: {
      display: 'inline-block'
    }
  };
};

var DisplayValueTypography =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(DisplayValueTypography, _React$PureComponent);

  function DisplayValueTypography() {
    (0, _classCallCheck2.default)(this, DisplayValueTypography);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(DisplayValueTypography).apply(this, arguments));
  }

  (0, _createClass2.default)(DisplayValueTypography, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          classes = _this$props.classes; // We use a span tag so that the element is inline

      return _react.default.createElement(_Typography.default, {
        variant: "subtitle1",
        component: "span",
        className: classes.root
      }, children);
    }
  }]);
  return DisplayValueTypography;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(DisplayValueTypography);

exports.default = _default;
