// mson-react/lib/fields/form-label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = _interopRequireDefault(require("react"));

var _FormLabel = _interopRequireDefault(require("@material-ui/core/FormLabel"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = function styles(theme) {
  return {
    shrink: (0, _objectSpread2.default)({}, Object.assign({}, theme.typography.caption, {
      color: undefined
    })),
    noShrink: (0, _objectSpread2.default)({}, Object.assign({}, theme.typography.subtitle1, {
      color: undefined
    }))
  };
};

var FormLabel =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(FormLabel, _React$PureComponent);

  function FormLabel() {
    (0, _classCallCheck2.default)(this, FormLabel);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FormLabel).apply(this, arguments));
  }

  (0, _createClass2.default)(FormLabel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          classes = _this$props.classes,
          shrink = _this$props.shrink;
      var childProps = Object.assign({}, this.props, {
        shrink: undefined,
        classes: undefined
      });
      var className = shrink ? classes.shrink : classes.noShrink;
      return _react.default.createElement(_FormLabel.default, (0, _extends2.default)({}, childProps, {
        className: className
      }), children);
    }
  }]);
  return FormLabel;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(FormLabel);

exports.default = _default;
