// mson-react/lib/fields/form-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _commonField = _interopRequireDefault(require("./common-field"));

var _component = _interopRequireDefault(require("../component"));

var _attach = _interopRequireDefault(require("../attach"));

var _flexBreak = _interopRequireDefault(require("../flex-break"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = function styles(theme) {
  return {
    root: {
      marginLeft: theme.spacing.unit * 2,
      // The following flex properties are needed so that we can set a margin for all fields in the
      // nested form
      display: 'flex',
      flexFlow: 'wrap'
    }
  };
};

var FormField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(FormField, _React$PureComponent);

  function FormField() {
    (0, _classCallCheck2.default)(this, FormField);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FormField).apply(this, arguments));
  }

  (0, _createClass2.default)(FormField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          component = _this$props.component,
          hideLabel = _this$props.hideLabel,
          classes = _this$props.classes,
          disabled = _this$props.disabled;
      var isBlank = component.isBlank();
      var form = component.get('form');
      return _react.default.createElement("span", null, !hideLabel && _react.default.createElement(_commonField.default, {
        component: component,
        inlineLabel: "true",
        shrinkLabel: !isBlank,
        marginBottom: false
      }), _react.default.createElement(_flexBreak.default, null), _react.default.createElement("span", {
        className: classes.root
      }, _react.default.createElement(_component.default, {
        component: form,
        formTag: false,
        disabled: disabled
      })));
    }
  }]);
  return FormField;
}(_react.default.PureComponent);

FormField = (0, _withStyles.default)(styles)(FormField);

var _default = (0, _attach.default)(['hideLabel', 'value' // Changes when the value is or isn't blank
])(FormField);

exports.default = _default;
