// mson-react/lib/fields/list-item-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _field = _interopRequireDefault(require("./field"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _icon = _interopRequireDefault(require("../icon"));

var _flexBreak = _interopRequireDefault(require("../flex-break"));

var _attach = _interopRequireDefault(require("../attach"));

var _helpToolTip = _interopRequireDefault(require("./help-tool-tip"));

var ListItemField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(ListItemField, _React$PureComponent);

  function ListItemField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ListItemField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ListItemField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDelete", function () {
      _this.props.component.emit('delete');
    });
    return _this;
  }

  (0, _createClass2.default)(ListItemField, [{
    key: "render",
    value: function render() {
      // Note: allowDelete and help have to be passed in as a prop as it is really the parent's values
      // that we need
      var _this$props = this.props,
          component = _this$props.component,
          allowDelete = _this$props.allowDelete,
          help = _this$props.help,
          disabled = _this$props.disabled,
          editable = _this$props.editable;

      if (component) {
        var block = component.get('block');
        return _react.default.createElement("span", null, _react.default.createElement(_field.default, {
          component: component,
          block: false,
          disabled: disabled
        }), allowDelete && !disabled && editable ? _react.default.createElement(_IconButton.default, {
          "aria-label": "Delete",
          onClick: this.handleDelete
        }, _react.default.createElement(_icon.default, {
          icon: "DeleteIcon"
        })) : null, help && editable ? _react.default.createElement(_helpToolTip.default, {
          help: help
        }) : '', block ? _react.default.createElement(_flexBreak.default, null) : null);
      } else {
        // component can be falsy if it was just deleteed
        return null;
      }
    }
  }]);
  return ListItemField;
}(_react.default.PureComponent); // We want the component to update when the value changes as blank fields should not display a
// delete button


var _default = (0, _attach.default)(['value', 'disabled', 'editable'])(ListItemField);

exports.default = _default;
