// mson-react/lib/fields/time-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("../attach"));

var _commonField = _interopRequireDefault(require("./common-field"));

var _MuiPickersUtilsProvider = _interopRequireDefault(require("material-ui-pickers/MuiPickersUtilsProvider"));

var _dateFns = _interopRequireDefault(require("@date-io/date-fns"));

var _TimePicker = _interopRequireDefault(require("material-ui-pickers/TimePicker"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _displayValueTypography = _interopRequireDefault(require("./display-value-typography"));

var styles = function styles(theme) {
  return {
    root: {
      // We need to add a margin as the picker doesn't play well with the Material-UI label
      marginTop: theme.spacing.unit * 2
    }
  };
};

var TimeField =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(TimeField, _React$PureComponent);

  function TimeField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, TimeField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(TimeField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDateChange", function (date) {
      _this.props.component.setValue(date);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function () {
      _this.props.component.setTouched(true);
    });
    return _this;
  }

  (0, _createClass2.default)(TimeField, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          component = _this$props.component,
          classes = _this$props.classes,
          value = _this$props.value,
          editable = _this$props.editable,
          useDisplayValue = _this$props.useDisplayValue,
          showSeconds = _this$props.showSeconds,
          fullWidth = _this$props.fullWidth,
          disabled = _this$props.disabled,
          accessEditable = _this$props.accessEditable;
      var dis = accessEditable === false || disabled;
      var shrinkLabel = false;
      var fld = null;

      if (editable && !useDisplayValue) {
        // The picker doesn't play well with the label from Material-UI so we need to manually shrink
        // the label when there is a value.
        shrinkLabel = !!value;
        fld = _react.default.createElement(_MuiPickersUtilsProvider.default, {
          utils: _dateFns.default
        }, _react.default.createElement("span", null, _react.default.createElement(_TimePicker.default // Value cannot be undefined or else picker will default to today
        , {
          value: value ? value : null,
          onChange: this.handleDateChange,
          clearable: true,
          className: classes.root,
          seconds: showSeconds // The minDate and maxDate are not used as they are confusing. They impose a
          // restriction on the actual day, which doesn't make sense as the time is store as a
          // timestamp with today's date and therefore the days may not match. Moreover, the
          // picker doesn't change when these dates are specified so we can just handle the
          // validation at the MSON-core layer. TODO: we can implement these ranges in the
          // MSON-core layer by constructing dates with the current day and specified time from
          // minDate and maxDate.
          // minDate={minDate}
          // maxDate={maxDate}
          ,
          onClose: this.handleBlur,
          fullWidth: fullWidth,
          disabled: dis,
          invalidDateMessage: "" // Let CommonField display the error
          // format="M/d/YYYY h:m a"

        })));
      } else {
        fld = _react.default.createElement(_displayValueTypography.default, null, component.getDisplayValue());
      }

      return _react.default.createElement(_commonField.default, {
        component: component,
        shrinkLabel: shrinkLabel
      }, fld);
    }
  }]);
  return TimeField;
}(_react.default.PureComponent);

TimeField = (0, _withStyles.default)(styles)(TimeField);

var _default = (0, _attach.default)(['value', 'editable', 'useDisplayValue', 'showSeconds', 'fullWidth', 'disabled'])(TimeField);

exports.default = _default;
