// mson/lib/fields/chained-select-list-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _listField = _interopRequireDefault(require("./list-field"));

var _chainedSelectField = _interopRequireDefault(require("./chained-select-field"));

// TODO: Should be able to covert this to a JSON field, right?
var ChainedSelectListField =
/*#__PURE__*/
function (_ListField) {
  (0, _inherits2.default)(ChainedSelectListField, _ListField);

  function ChainedSelectListField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ChainedSelectListField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ChainedSelectListField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'ChainedSelectListField');
    return _this;
  }

  (0, _createClass2.default)(ChainedSelectListField, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(ChainedSelectListField.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'fieldFactory',
            required: false
          }, {
            name: 'options',
            component: 'ChainedSelectOptionsField',
            docLevel: 'basic'
          }, {
            name: 'blankString',
            component: 'TextField'
          }]
        }
      });

      this._setDefaults(props, {
        allowDelete: true,
        autoCreateFields: true,
        canDeleteEmpty: false,
        startWithField: true
      });
    }
  }, {
    key: "_newField",
    value: function _newField(name) {
      return new _chainedSelectField.default((0, _objectSpread2.default)({
        name: name,
        hideLabel: true,
        block: true
      }, this.get(['blankString', 'fullWidth', 'options', 'useDisplayValue', 'editable'])));
    }
  }, {
    key: "_setFieldOptions",
    value: function _setFieldOptions(name) {
      var field = this._getOrCreateField(name);

      field.set({
        options: this.get('options')
      });
    }
  }, {
    key: "_onFieldCreated",
    value: function _onFieldCreated(field
    /* onDelete */
    ) {
      var _this2 = this;

      field.on('value', function (value) {
        // Create last field if the last field is not blank and we are allowed to add more fields
        if (!_this2._isLastFieldBlank() && _this2.canAddMoreFields(value)) {
          _this2.createField();
        }

        _this2._calcValue();
      });
    }
  }, {
    key: "_setOptions",
    value: function _setOptions(options) {
      // Set the options for the first field
      this._setFieldOptions(0); // Set options for all fields


      this.eachField(function (field) {
        return field.set({
          options: options
        });
      });
    }
  }, {
    key: "set",
    value: function set(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(ChainedSelectListField.prototype), "set", this).call(this, props);

      if (props.options !== undefined) {
        this._setOptions(props.options);
      }
    }
  }, {
    key: "clone",
    value: function clone() {
      var clonedField = (0, _get2.default)((0, _getPrototypeOf3.default)(ChainedSelectListField.prototype), "clone", this).call(this); // Clear the fields as the listeners now have the wrong references to the fields, etc...

      clonedField._fields.clear(); // Create the first field


      clonedField._setOptions(this.get('options')); // Copy any existing value


      clonedField.setValue(this.getValue());
      return clonedField;
    }
  }]);
  return ChainedSelectListField;
}(_listField.default);

exports.default = ChainedSelectListField;
