// mson/lib/fields/composite-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _field = _interopRequireDefault(require("./field"));

var _mapa = _interopRequireDefault(require("../mapa"));

var _clone = _interopRequireDefault(require("lodash/clone"));

var _isEmpty = _interopRequireDefault(require("lodash/isEmpty"));

var CompositeField =
/*#__PURE__*/
function (_Field) {
  (0, _inherits2.default)(CompositeField, _Field);

  function CompositeField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, CompositeField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(CompositeField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'CompositeField');
    return _this;
  }

  (0, _createClass2.default)(CompositeField, [{
    key: "_create",
    value: function _create(props) {
      // We use a Mapa instead of an array as sometimes we need to reference the fields via keys that
      // don't change even when fields are deleted. With arrays, the keys change when we slice the
      // array. We cannot use just a basic object as our fields must preseve order. We cannot use a
      // Map as we need to be able to iterate through the fields beginning with any given field.
      this._fields = new _mapa.default();
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'fields',
            component: 'Field'
          }, {
            name: 'change',
            component: 'Field'
          }]
        }
      });
    }
  }, {
    key: "_bubbleUpTouches",
    value: function _bubbleUpTouches(field) {
      var _this2 = this;

      // Bubble up the touched events from the children
      field.on('touched', function (touched) {
        if (touched) {
          var allTouched = true;

          _this2.eachField(function (field) {
            if (!field.get('touched')) {
              allTouched = false;
              return false; // exit loop
            }
          });

          if (allTouched) {
            // The parent field is not considered touched until all the child fields have been
            // touched. This prevents prematurely showing errors like "required" errors when the user
            // tabs from first field.
            //
            // We use _set() instead of setTouched as we only want to set the touched
            // property at this layer and not for the sub fields.
            _this2._set('touched', touched);
          }
        }
      });
    }
  }, {
    key: "_setParentValue",
    value: function _setParentValue(field, value) {
      // Clone the value as we don't want to modify this._value directly as we want set to be able
      // to detect if the value is changing.
      var clonedValue = !this._value ? {} : (0, _clone.default)(this._value);

      if (value === null) {
        delete clonedValue[field.get('name')];
      } else {
        clonedValue[field.get('name')] = value;
      }

      this.setValue((0, _isEmpty.default)(clonedValue) ? null : clonedValue);
    }
  }, {
    key: "_listenForChangesToField",
    value: function _listenForChangesToField(field) {
      var _this3 = this;

      // Merge the value changes from the children into the parent value
      field.on('value', function (value) {
        return _this3._setParentValue(field, value);
      }); // Field may already have a value

      this._setParentValue(field, field.getValue());

      this._bubbleUpTouches(field);
    }
  }, {
    key: "emitChangeToField",
    value: function emitChangeToField(field) {
      // Emit event so that we do things like dynamically adjust the display of fields. We can't emit
      // just all the fields as otherwise a shallow comparison won't detect a change. Also, we have to
      // use field.get() so that a shallow comparison will detect a difference in the fields.
      // this.set({ change: field });
      this.set({
        change: field.get()
      });
    }
  }, {
    key: "_addField",
    value: function _addField(field, name) {
      if (name === undefined) {
        name = field.get('name');
      }

      this._fields.set(name, field);

      this.emitChangeToField(field);

      this._listenForChangesToField(field);
    }
  }, {
    key: "_setFields",
    value: function _setFields(fields) {
      var _this4 = this;

      fields.forEach(function (field) {
        return _this4._addField(field);
      });
    }
  }, {
    key: "_setForAllFields",
    value: function _setForAllFields(props, names) {
      var propsToSet = null;

      if (names) {
        propsToSet = {};
        names.forEach(function (name) {
          if (props[name] !== undefined) {
            propsToSet[name] = props[name];
          }
        });
      } else {
        propsToSet = props;
      }

      if (!(0, _isEmpty.default)(propsToSet)) {
        this.eachField(function (field) {
          return field.set(propsToSet);
        });
      }
    }
  }, {
    key: "_setValue",
    value: function _setValue(value) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_setValue", this).call(this, value);
      this.eachField(function (field) {
        return field.setValue(value && value[field.get('name')] ? value[field.get('name')] : null);
      });
    }
  }, {
    key: "_setRequired",
    value: function _setRequired(required) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_setRequired", this).call(this, required);

      this._setForAllFields({
        required: required
      });
    }
  }, {
    key: "_setDisabled",
    value: function _setDisabled(disabled) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_setDisabled", this).call(this, disabled);

      this._setForAllFields({
        disabled: disabled
      });
    }
  }, {
    key: "_setEditable",
    value: function _setEditable(editable) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_setEditable", this).call(this, editable);

      this._setForAllFields({
        editable: editable
      });
    }
  }, {
    key: "_setDirty",
    value: function _setDirty(dirty) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "_setDirty", this).call(this, dirty);

      this._setForAllFields({
        dirty: dirty
      });
    }
  }, {
    key: "_setTouchedForAllFields",
    value: function _setTouchedForAllFields(touched) {
      this._setForAllFields({
        touched: touched
      });
    }
  }, {
    key: "setTouched",
    value: function setTouched(touched) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "setTouched", this).call(this, touched);

      this._setTouchedForAllFields(touched);
    }
  }, {
    key: "setFullWidth",
    value: function setFullWidth(fullWidth) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "setFullWidth", this).call(this, fullWidth);

      this._setForAllFields({
        fullWidth: fullWidth
      });
    }
  }, {
    key: "setUseDisplayValue",
    value: function setUseDisplayValue(useDisplayValue) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "setUseDisplayValue", this).call(this, useDisplayValue);

      this._setForAllFields({
        useDisplayValue: useDisplayValue
      });
    }
  }, {
    key: "setHideLabel",
    value: function setHideLabel(hideLabel) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "setHideLabel", this).call(this, hideLabel);

      this._setForAllFields({
        hideLabel: hideLabel
      });
    }
  }, {
    key: "set",
    value: function set(props) {
      var _this5 = this;

      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "set", this).call(this, Object.assign({}, props, {
        fields: undefined
      }));

      if (props.fields !== undefined) {
        this._setFields(props.fields);
      }

      this.eachField(function (field) {
        var name = field.get('name');

        _this5._setIfDefined(name, props[name]);
      });
    }
  }, {
    key: "getField",
    value: function getField(name) {
      return this._fields.get(name);
    }
  }, {
    key: "_getFieldIfExists",
    value: function _getFieldIfExists(name) {
      if (this._fields.has(name)) {
        return this.getField(name);
      }
    }
  }, {
    key: "has",
    value: function has(name) {
      return (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "has", this).call(this, name) || this._fields.has(name);
    }
  }, {
    key: "getOne",
    value: function getOne(name) {
      var value = this._getFieldIfExists(name);

      return value === undefined ? (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "getOne", this).call(this, name) : value;
    }
  }, {
    key: "eachField",
    value: function eachField(onField) {
      this._fields.each(function (field, name, last) {
        return onField(field, name, last);
      });
    }
  }, {
    key: "mapFields",
    value: function mapFields(onField) {
      return this._fields.map(function (field, name, last) {
        return onField(field, name, last);
      });
    }
  }, {
    key: "clearErr",
    value: function clearErr() {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "clearErr", this).call(this);
      this.eachField(function (field) {
        return field.clearErr();
      });
    }
  }, {
    key: "_validateField",
    value: function _validateField(field
    /*, name, last */
    ) {
      field.validate();
    }
  }, {
    key: "validate",
    value: function validate() {
      var _this6 = this;

      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "validate", this).call(this);
      this.eachField(function (field, name, last) {
        return _this6._validateField(field, name, last);
      });
    }
  }, {
    key: "getDisplayValue",
    value: function getDisplayValue() {
      if (this.isBlank()) {
        return null;
      } else {
        return this.mapFields(function (field) {
          return field.getDisplayValue();
        });
      }
    }
  }, {
    key: "isBlank",
    value: function isBlank() {
      if ((0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "isBlank", this).call(this)) {
        return true;
      } else {
        var value = this.getValue();
        return Array.isArray(value) && value.length === 0;
      }
    }
  }, {
    key: "destroy",
    value: function destroy() {
      (0, _get2.default)((0, _getPrototypeOf3.default)(CompositeField.prototype), "destroy", this).call(this);
      this.eachField(function (field) {
        return field.destroy();
      });
    }
  }]);
  return CompositeField;
}(_field.default);

exports.default = CompositeField;
