// mson/lib/fields/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ButtonField", {
  enumerable: true,
  get: function get() {
    return _buttonField.default;
  }
});
Object.defineProperty(exports, "DateField", {
  enumerable: true,
  get: function get() {
    return _dateField.default;
  }
});
Object.defineProperty(exports, "Field", {
  enumerable: true,
  get: function get() {
    return _field.default;
  }
});
Object.defineProperty(exports, "FormField", {
  enumerable: true,
  get: function get() {
    return _formField.default;
  }
});
Object.defineProperty(exports, "CollectionField", {
  enumerable: true,
  get: function get() {
    return _collectionField.default;
  }
});
Object.defineProperty(exports, "IdField", {
  enumerable: true,
  get: function get() {
    return _idField.default;
  }
});
Object.defineProperty(exports, "IntegerField", {
  enumerable: true,
  get: function get() {
    return _integerField.default;
  }
});
Object.defineProperty(exports, "ListField", {
  enumerable: true,
  get: function get() {
    return _listField.default;
  }
});
Object.defineProperty(exports, "NumberField", {
  enumerable: true,
  get: function get() {
    return _numberField.default;
  }
});
Object.defineProperty(exports, "SelectField", {
  enumerable: true,
  get: function get() {
    return _selectField.default;
  }
});
Object.defineProperty(exports, "TextField", {
  enumerable: true,
  get: function get() {
    return _textField.default;
  }
});
Object.defineProperty(exports, "TextListField", {
  enumerable: true,
  get: function get() {
    return _textListField.default;
  }
});
exports.PersonNameField = exports.PersonFullNameField = exports.PasswordField = exports.EmailField = void 0;

var _compiler = _interopRequireDefault(require("../compiler"));

var _buttonField = _interopRequireDefault(require("./button-field"));

var _dateField = _interopRequireDefault(require("./date-field"));

var _field = _interopRequireDefault(require("./field"));

var _formField = _interopRequireDefault(require("./form-field"));

var _collectionField = _interopRequireDefault(require("./collection-field"));

var _idField = _interopRequireDefault(require("./id-field"));

var _integerField = _interopRequireDefault(require("./integer-field"));

var _listField = _interopRequireDefault(require("./list-field"));

var _numberField = _interopRequireDefault(require("./number-field"));

var _selectField = _interopRequireDefault(require("./select-field"));

var _textField = _interopRequireDefault(require("./text-field"));

var _textListField = _interopRequireDefault(require("./text-list-field"));

// For readability
var compile = _compiler.default.getCompiledComponent.bind(_compiler.default); // Compile the uncompiled components so that they can be used like any other compiled component.
// This adds a little processing, but it is quick because the compiler just creates a new class that
// wraps the MSON, which means that the bulk of the work is deferred until the component is
// instantiated.


var EmailField = compile('EmailField');
exports.EmailField = EmailField;
var PersonFullNameField = compile('PersonFullNameField');
exports.PersonFullNameField = PersonFullNameField;
var PersonNameField = compile('PersonNameField');
exports.PersonNameField = PersonNameField;
var PasswordField = compile('PasswordField');
exports.PasswordField = PasswordField;
