// mson/lib/fields/text-list-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _listField = _interopRequireDefault(require("./list-field"));

var _textField = _interopRequireDefault(require("./text-field"));

var TextListField =
/*#__PURE__*/
function (_ListField) {
  (0, _inherits2.default)(TextListField, _ListField);

  function TextListField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, TextListField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(TextListField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'TextListField');
    return _this;
  }

  (0, _createClass2.default)(TextListField, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(TextListField.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'fieldFactory',
            required: false
          }, {
            name: 'minLength',
            component: 'IntegerField',
            label: 'Min Length',
            docLevel: 'basic'
          }, {
            name: 'maxLength',
            component: 'IntegerField',
            label: 'Max Length',
            docLevel: 'basic'
          }, {
            name: 'minWords',
            component: 'IntegerField',
            label: 'Min Words',
            docLevel: 'basic'
          }, {
            name: 'maxWords',
            component: 'IntegerField',
            label: 'Max Words',
            docLevel: 'basic'
          }, {
            name: 'invalidRegExp',
            component: 'TextField',
            label: 'Invalid RegExp',
            docLevel: 'basic'
          }, {
            name: 'multiline',
            component: 'BooleanField',
            label: 'Multiline',
            docLevel: 'basic'
          }]
        }
      });

      this._setDefaults(props, {
        startWithField: true
      });
    }
  }, {
    key: "_newField",
    value: function _newField(index) {
      return new _textField.default((0, _objectSpread2.default)({
        name: index,
        required: false
      }, this.get(['block', 'fullWidth', 'minLength', 'maxLength', 'minWords', 'maxWords', 'invalidRegExp', 'multiline'])));
    }
  }, {
    key: "set",
    value: function set(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(TextListField.prototype), "set", this).call(this, props);

      this._setForAllFields(props, ['minLength', 'maxLength', 'minWords', 'maxWords', 'invalidRegExp', 'multiline']);
    }
  }]);
  return TextListField;
}(_listField.default);

exports.default = TextListField;
