// mson/lib/fields/time-field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _field = _interopRequireDefault(require("./field"));

var _dateField = _interopRequireDefault(require("./date-field"));

var _isValid = _interopRequireDefault(require("date-fns/isValid"));

var TimeField =
/*#__PURE__*/
function (_Field) {
  (0, _inherits2.default)(TimeField, _Field);

  function TimeField() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, TimeField);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(TimeField)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'TimeField');
    return _this;
  }

  (0, _createClass2.default)(TimeField, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(TimeField.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'showSeconds',
            component: 'BooleanField',
            label: 'Show Seconds',
            docLevel: 'basic'
          }]
        }
      });
    }
  }, {
    key: "set",
    value: function set(props) {
      var clonedProps = Object.assign({}, props); // Convert Date? We store dates in epoch time so that they are compatiable across all stores.
      // Epoch time is also smaller than the ISO string and can therefore minimize the storage needed.

      if (clonedProps.value !== undefined && typeof clonedProps.value !== 'number') {
        clonedProps.value = _dateField.default.toEpochTime(clonedProps.value);
      }

      (0, _get2.default)((0, _getPrototypeOf3.default)(TimeField.prototype), "set", this).call(this, clonedProps);
    } // For mocking

  }, {
    key: "_toLocaleString",
    value: function _toLocaleString(date) {
      var currentLocale = [];
      var options = {};

      if (!this.get('showSeconds')) {
        options = {
          hour: '2-digit',
          minute: '2-digit'
        };
      }

      return date.toLocaleTimeString(currentLocale, options);
    }
  }, {
    key: "getDisplayValue",
    value: function getDisplayValue() {
      return this.isBlank() ? null : this._toLocaleString(new Date(this.getValue()));
    }
  }, {
    key: "validate",
    value: function validate() {
      (0, _get2.default)((0, _getPrototypeOf3.default)(TimeField.prototype), "validate", this).call(this);

      if (!this.isBlank()) {
        var value = this.getValue();
        var date = new Date(value);

        if (!(0, _isValid.default)(date)) {
          this.setErr('invalid');
        }
      }
    }
  }]);
  return TimeField;
}(_field.default);

exports.default = TimeField;
