// mson/lib/actions/console-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _action = _interopRequireDefault(require("./action"));

// Can be used to debug MSON component actions
var ConsoleLog =
/*#__PURE__*/
function (_Action) {
  (0, _inherits2.default)(ConsoleLog, _Action);

  function ConsoleLog() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ConsoleLog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ConsoleLog)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'ConsoleLog');
    return _this;
  }

  (0, _createClass2.default)(ConsoleLog, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(ConsoleLog.prototype), "_create", this).call(this, props); // For mocking

      this._console = console;
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'message',
            component: 'TextField'
          }]
        }
      });
    }
  }, {
    key: "act",
    value: function () {
      var _act = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(props) {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                this._console.log(this.get('message')); // Pass on the arguments so that the log can be used in the middle of 2 components


                return _context.abrupt("return", props.arguments);

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function act(_x) {
        return _act.apply(this, arguments);
      }

      return act;
    }()
  }]);
  return ConsoleLog;
}(_action.default);

exports.default = ConsoleLog;
