// mson/lib/actions/iterator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _action = _interopRequireDefault(require("./action"));

var _get3 = _interopRequireDefault(require("lodash/get"));

var _clone = _interopRequireDefault(require("lodash/clone"));

var _map = _interopRequireDefault(require("lodash/map"));

var _propFiller = _interopRequireDefault(require("../compiler/prop-filler"));

var Iterator =
/*#__PURE__*/
function (_Action) {
  (0, _inherits2.default)(Iterator, _Action);

  function Iterator() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Iterator);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Iterator)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'Iterator');
    return _this;
  }

  (0, _createClass2.default)(Iterator, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(Iterator.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'iterator',
            component: 'Field',
            required: true
          }, {
            name: 'return',
            component: 'Field',
            required: true
          }]
        }
      });
    }
  }, {
    key: "_getProp",
    value: function _getProp(props, path) {
      return (0, _get3.default)(props, path);
    }
  }, {
    key: "act",
    value: function () {
      var _act = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(props) {
        var _this2 = this;

        var iterator, clonedProps, filler;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                iterator = this._getProp(props, this.get('iterator'));
                clonedProps = (0, _clone.default)(props);
                filler = new _propFiller.default(props);
                return _context.abrupt("return", (0, _map.default)(iterator, function (item) {
                  // Inject item
                  clonedProps.item = item;
                  filler.setProps(clonedProps);
                  return filler.fillAll(_this2.get('return'));
                }));

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function act(_x) {
        return _act.apply(this, arguments);
      }

      return act;
    }()
  }]);
  return Iterator;
}(_action.default);

exports.default = Iterator;
