// mson/lib/actions/redirect-by-role.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _action = _interopRequireDefault(require("./action"));

var _each = _interopRequireDefault(require("lodash/each"));

var RedirectByRole =
/*#__PURE__*/
function (_Action) {
  (0, _inherits2.default)(RedirectByRole, _Action);

  function RedirectByRole() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, RedirectByRole);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(RedirectByRole)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'RedirectByRole');
    return _this;
  }

  (0, _createClass2.default)(RedirectByRole, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(RedirectByRole.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'routes',
            component: 'Field',
            required: true
          }]
        }
      });

      this._setDefaults(props, {
        layer: 'frontEnd'
      });
    }
  }, {
    key: "act",
    value: function () {
      var _act = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee() {
        var _this2 = this;

        var routes;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                routes = this.get('routes');
                (0, _each.default)(routes, function (route) {
                  if (!route.roles || _this2._registrar.client.user.hasRole(route.roles)) {
                    _this2._globals.redirect(route.path);

                    return false; // exit loop
                  }
                });

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function act() {
        return _act.apply(this, arguments);
      }

      return act;
    }()
  }]);
  return RedirectByRole;
}(_action.default);

exports.default = RedirectByRole;
