// mson/lib/actions/set.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _action = _interopRequireDefault(require("./action"));

var Set =
/*#__PURE__*/
function (_Action) {
  (0, _inherits2.default)(Set, _Action);

  function Set() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Set);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Set)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'Set');
    return _this;
  }

  (0, _createClass2.default)(Set, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(Set.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'name',
            component: 'TextField'
          }, {
            name: 'value',
            component: 'Field'
          }]
        }
      });
    }
  }, {
    key: "_setProp",
    value: function _setProp(props) {
      var name = this.get('name');
      var names = name ? name.split('.') : [];
      var thisValue = this.get('value');
      var value = thisValue === undefined ? props.arguments : thisValue;

      if (!name) {
        // No name was specified, pipe to next action
        return value;
      } else if (names[0] === 'globals' || names[0] === 'component') {
        var first = names.splice(0, 1)[0]; // Remove globals or component

        if (first === 'globals') {
          this._globals.set((0, _defineProperty2.default)({}, names.join('.'), value));
        } else {
          props.component.set(value);
        }
      } else {
        props.component.set((0, _defineProperty2.default)({}, name, value));
      } // Pipe the arguments so that we can do things like use multiple Set actions to copy pieces of
      // our API results


      return props.arguments;
    }
  }, {
    key: "act",
    value: function () {
      var _act = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(props) {
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", this._setProp(props));

              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function act(_x) {
        return _act.apply(this, arguments);
      }

      return act;
    }()
  }]);
  return Set;
}(_action.default);

exports.default = Set;
