// mson/lib/stores/firebase-mock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _mapa = _interopRequireDefault(require("../mapa"));

// Mock Firebase so that we don't have to test against the actual service
var SnapshotMock =
/*#__PURE__*/
function () {
  function SnapshotMock(docs) {
    (0, _classCallCheck2.default)(this, SnapshotMock);
    this._docs = docs;
  }

  (0, _createClass2.default)(SnapshotMock, [{
    key: "docChanges",
    value: function docChanges() {
      return this._docs.map(function (doc) {
        return {
          type: doc.type,
          doc: {
            data: function data() {
              return doc.data;
            }
          }
        };
      });
    }
  }]);
  return SnapshotMock;
}();

var CollectionMock =
/*#__PURE__*/
function () {
  (0, _createClass2.default)(CollectionMock, [{
    key: "_listenToChanges",
    value: function _listenToChanges() {// Not currently needed by our tests, but may be used in the future.
      //
      // Simulate the response from Firebase after a write
      // this._docs.on('$change', (name, value) => {
      //   switch (name) {
      //     case 'createDoc':
      //       this.emitSnapshot([ { type: 'added', data: value.value } ])
      //       break;
      //     case 'updateDoc':
      //       this.emitSnapshot([ { type: 'modified', data: value.value } ])
      //       break;
      //     case 'deleteDoc':
      //       this.emitSnapshot([ { type: 'removed', data: value.value } ])
      //       break;
      //     default:
      //       break;
      //   }
      // });
    }
  }]);

  function CollectionMock(docs) {
    (0, _classCallCheck2.default)(this, CollectionMock);
    this._docs = docs;

    this._listenToChanges();
  }

  (0, _createClass2.default)(CollectionMock, [{
    key: "orderBy",
    value: function orderBy() {
      return this;
    }
  }, {
    key: "emitSnapshot",
    value: function emitSnapshot(docs) {
      this._onSnapshot(new SnapshotMock(docs));
    }
  }, {
    key: "emitError",
    value: function emitError(err) {
      this._onError(err);
    }
  }, {
    key: "onSnapshot",
    value: function onSnapshot(_onSnapshot, onError) {
      this._onSnapshot = _onSnapshot;
      this._onError = onError; // Emit initial data

      this.emitSnapshot(this._docs.map(function (doc) {
        return {
          type: 'added',
          data: doc.value
        };
      }));
    }
  }, {
    key: "doc",
    value: function doc(id) {
      var _this = this;

      return {
        set: function () {
          var _set = (0, _asyncToGenerator2.default)(
          /*#__PURE__*/
          _regenerator.default.mark(function _callee(doc) {
            return _regenerator.default.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    return _context.abrupt("return", _this._docs.set(id, doc));

                  case 1:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          }));

          function set(_x) {
            return _set.apply(this, arguments);
          }

          return set;
        }()
      };
    }
  }]);
  return CollectionMock;
}();

var FirebaseMock =
/*#__PURE__*/
function () {
  function FirebaseMock(docs) {
    var _this2 = this;

    (0, _classCallCheck2.default)(this, FirebaseMock);
    this._docs = new _mapa.default();

    if (docs) {
      docs.forEach(function (doc) {
        return _this2._docs.set(doc.id, doc);
      });
    }

    this._collection = new CollectionMock(this._docs);
    this.apps = {
      length: 0
    };
  }

  (0, _createClass2.default)(FirebaseMock, [{
    key: "initializeApp",
    value: function initializeApp() {
      this.apps.length = 1;
    }
  }, {
    key: "firestore",
    value: function firestore() {
      var _this3 = this;

      return {
        settings: function settings() {},
        collection: function collection() {
          return _this3._collection;
        }
      };
    }
  }, {
    key: "emitSnapshot",
    value: function emitSnapshot(docs) {
      this._collection.emitSnapshot(docs);
    }
  }, {
    key: "emitError",
    value: function emitError(err) {
      this._collection.emitError(err);
    }
  }]);
  return FirebaseMock;
}();

exports.default = FirebaseMock;
