// mson/lib/stores/local-storage-store.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _memoryStore = _interopRequireDefault(require("./memory-store"));

var LocalStorageStore =
/*#__PURE__*/
function (_MemoryStore) {
  (0, _inherits2.default)(LocalStorageStore, _MemoryStore);

  function LocalStorageStore() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, LocalStorageStore);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(LocalStorageStore)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'LocalStorageStore');
    return _this;
  }

  (0, _createClass2.default)(LocalStorageStore, [{
    key: "_loadDocs",
    value: function _loadDocs(storeName) {
      var _this2 = this;

      var store = this._localStorage.getItem(storeName);

      if (store !== null) {
        var docs = JSON.parse(store);
        docs.forEach(function (doc) {
          _this2._docs.set(doc.id, doc);
        });
      }
    } // For mocking

  }, {
    key: "_getLocalStorage",
    value: function _getLocalStorage() {
      return this._global.window.localStorage;
    }
  }, {
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(LocalStorageStore.prototype), "_create", this).call(this, props); // For mocking

      this._global = global;
      this._localStorage = this._getLocalStorage();
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'storeName',
            component: 'TextField',
            required: true
          }]
        }
      });

      this._loadDocs(props && props.storeName);
    }
  }, {
    key: "_saveDocs",
    value: function _saveDocs() {
      var docs = this._docs.map(function (doc) {
        return doc;
      });

      this._localStorage.setItem(this.get('storeName'), JSON.stringify(docs));
    }
  }, {
    key: "_doAndSave",
    value: function () {
      var _doAndSave2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(promiseFactory) {
        var response;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return promiseFactory();

              case 2:
                response = _context.sent;

                this._saveDocs();

                return _context.abrupt("return", response);

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function _doAndSave(_x) {
        return _doAndSave2.apply(this, arguments);
      }

      return _doAndSave;
    }()
  }, {
    key: "_createDoc",
    value: function () {
      var _createDoc2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(props) {
        var _this3 = this;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", this._doAndSave(function () {
                  return (0, _get2.default)((0, _getPrototypeOf3.default)(LocalStorageStore.prototype), "_createDoc", _this3).call(_this3, props);
                }));

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function _createDoc(_x2) {
        return _createDoc2.apply(this, arguments);
      }

      return _createDoc;
    }()
  }, {
    key: "_updateDoc",
    value: function () {
      var _updateDoc2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(props) {
        var _this4 = this;

        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                return _context3.abrupt("return", this._doAndSave(function () {
                  return (0, _get2.default)((0, _getPrototypeOf3.default)(LocalStorageStore.prototype), "_updateDoc", _this4).call(_this4, props);
                }));

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function _updateDoc(_x3) {
        return _updateDoc2.apply(this, arguments);
      }

      return _updateDoc;
    }()
  }, {
    key: "_archiveDoc",
    value: function () {
      var _archiveDoc2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(props) {
        var _this5 = this;

        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                return _context4.abrupt("return", this._doAndSave(function () {
                  return (0, _get2.default)((0, _getPrototypeOf3.default)(LocalStorageStore.prototype), "_archiveDoc", _this5).call(_this5, props);
                }));

              case 1:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function _archiveDoc(_x4) {
        return _archiveDoc2.apply(this, arguments);
      }

      return _archiveDoc;
    }()
  }, {
    key: "_restoreDoc",
    value: function () {
      var _restoreDoc2 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee5(props) {
        var _this6 = this;

        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                return _context5.abrupt("return", this._doAndSave(function () {
                  return (0, _get2.default)((0, _getPrototypeOf3.default)(LocalStorageStore.prototype), "_restoreDoc", _this6).call(_this6, props);
                }));

              case 1:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function _restoreDoc(_x5) {
        return _restoreDoc2.apply(this, arguments);
      }

      return _restoreDoc;
    }()
  }]);
  return LocalStorageStore;
}(_memoryStore.default);

exports.default = LocalStorageStore;
