// mson/lib/stores/record-mock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _memoryStore = _interopRequireDefault(require("./memory-store"));

// Mock the record API so that we don't need to test against an actual server
var RecordMock =
/*#__PURE__*/
function () {
  function RecordMock() {
    (0, _classCallCheck2.default)(this, RecordMock);
    // We use a MemoryStore that we get automatic sorting, querying, etc...
    this._docs = new _memoryStore.default();
  }

  (0, _createClass2.default)(RecordMock, [{
    key: "create",
    value: function () {
      var _create = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(_ref) {
        var appId, componentName, fieldValues, order, doc;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                appId = _ref.appId, componentName = _ref.componentName, fieldValues = _ref.fieldValues, order = _ref.order;
                _context.next = 3;
                return this._docs._createDoc({
                  fieldValues: fieldValues,
                  order: order
                });

              case 3:
                doc = _context.sent;
                return _context.abrupt("return", {
                  data: {
                    createRecord: doc
                  }
                });

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function create(_x) {
        return _create.apply(this, arguments);
      }

      return create;
    }()
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(_ref2) {
        var appId, componentName, id, where, doc;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                appId = _ref2.appId, componentName = _ref2.componentName, id = _ref2.id, where = _ref2.where;
                _context2.prev = 1;
                _context2.next = 4;
                return this._docs.getDoc({
                  id: id,
                  where: where
                });

              case 4:
                doc = _context2.sent;
                return _context2.abrupt("return", {
                  data: {
                    record: doc
                  }
                });

              case 8:
                _context2.prev = 8;
                _context2.t0 = _context2["catch"](1);
                throw new Error('GraphQL error: record not found');

              case 11:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[1, 8]]);
      }));

      function get(_x2) {
        return _get.apply(this, arguments);
      }

      return get;
    }()
  }, {
    key: "update",
    value: function () {
      var _update = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(_ref3) {
        var appId, componentName, id, fieldValues, order, doc;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                appId = _ref3.appId, componentName = _ref3.componentName, id = _ref3.id, fieldValues = _ref3.fieldValues, order = _ref3.order;
                _context3.next = 3;
                return this._docs._updateDoc({
                  id: id,
                  fieldValues: fieldValues,
                  order: order
                });

              case 3:
                doc = _context3.sent;
                return _context3.abrupt("return", {
                  data: {
                    updateRecord: doc
                  }
                });

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function update(_x3) {
        return _update.apply(this, arguments);
      }

      return update;
    }()
  }, {
    key: "archive",
    value: function () {
      var _archive = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(_ref4) {
        var appId, componentName, id, doc;
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                appId = _ref4.appId, componentName = _ref4.componentName, id = _ref4.id;
                _context4.next = 3;
                return this._docs.archiveDoc({
                  id: id
                });

              case 3:
                doc = _context4.sent;
                return _context4.abrupt("return", {
                  data: {
                    archiveRecord: doc
                  }
                });

              case 5:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function archive(_x4) {
        return _archive.apply(this, arguments);
      }

      return archive;
    }()
  }, {
    key: "restore",
    value: function () {
      var _restore = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee5(_ref5) {
        var appId, componentName, id, doc;
        return _regenerator.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                appId = _ref5.appId, componentName = _ref5.componentName, id = _ref5.id;
                _context5.next = 3;
                return this._docs.restoreDoc({
                  id: id
                });

              case 3:
                doc = _context5.sent;
                return _context5.abrupt("return", {
                  data: {
                    restoreRecord: doc
                  }
                });

              case 5:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function restore(_x5) {
        return _restore.apply(this, arguments);
      }

      return restore;
    }()
  }, {
    key: "getAll",
    value: function () {
      var _getAll = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee6(opts) {
        var docs;
        return _regenerator.default.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._docs.getAllDocs(opts);

              case 2:
                docs = _context6.sent;
                return _context6.abrupt("return", {
                  data: {
                    records: {
                      edges: docs.edges,
                      // TODO: this will probably need adjusted once our tests support pagination
                      pageInfo: {
                        hasNextPage: false
                      }
                    }
                  }
                });

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getAll(_x6) {
        return _getAll.apply(this, arguments);
      }

      return getAll;
    }()
  }]);
  return RecordMock;
}();

exports.default = RecordMock;
