// mson/lib/stores/store-test.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shouldMove = exports.shouldGetAll = exports.updateDoc = exports.createDoc = exports.shouldCRUD = exports.createForm = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _form = _interopRequireDefault(require("../form"));

var _fields = require("../fields");

var _testUtils = _interopRequireDefault(require("../test-utils"));

var _reorder = require("./reorder");

var createForm = function createForm(props) {
  return new _form.default((0, _objectSpread2.default)({
    fields: [new _fields.TextField({
      name: 'firstName'
    }), new _fields.TextField({
      name: 'lastName'
    })]
  }, props));
};

exports.createForm = createForm;

var shouldCRUD =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(Store, props) {
    var fieldValues, form, store, created, updated, archived, restored;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            fieldValues = {
              firstName: 'Ella',
              lastName: 'Fitzgerald'
            };
            form = createForm({
              value: fieldValues
            });
            store = new Store(props); // Create

            _context.next = 5;
            return store.createDoc({
              form: form
            });

          case 5:
            created = _context.sent;
            expect(created.id).not.toBeFalsy();
            expect(created.archivedAt).toBeNull();
            expect(created.createdAt).not.toBeFalsy();
            expect(created.updatedAt).not.toBeFalsy();
            expect(created.fieldValues).toEqual(fieldValues);
            expect(created.order).toEqual(_reorder.Reorder.DEFAULT_ORDER); // Get

            _context.t0 = expect;
            _context.next = 15;
            return store.getDoc({
              id: created.id
            });

          case 15:
            _context.t1 = _context.sent;
            _context.t2 = created;
            (0, _context.t0)(_context.t1).toEqual(_context.t2);
            _context.t3 = expect;
            _context.next = 21;
            return store.getDoc({
              where: {
                id: 'missingId'
              }
            });

          case 21:
            _context.t4 = _context.sent;
            (0, _context.t3)(_context.t4).toBeNull();
            form.setValues({
              id: created.id,
              firstName: 'F. Scott'
            }); // Make sure timestamps aren't the same

            _context.next = 26;
            return _testUtils.default.sleepToEnsureDifferentTimestamps();

          case 26:
            _context.next = 28;
            return store.updateDoc({
              id: created.id,
              form: form
            });

          case 28:
            updated = _context.sent;
            expect(updated).toEqual({
              id: created.id,
              archivedAt: null,
              createdAt: created.createdAt,
              updatedAt: updated.updatedAt,
              userId: null,
              fieldValues: {
                firstName: 'F. Scott',
                lastName: 'Fitzgerald'
              },
              order: _reorder.Reorder.DEFAULT_ORDER
            });
            expect(updated.updatedAt).not.toEqual(created.updatedAt);
            _context.t5 = expect;
            _context.next = 34;
            return store.getDoc({
              id: created.id
            });

          case 34:
            _context.t6 = _context.sent;
            _context.t7 = updated;
            (0, _context.t5)(_context.t6).toEqual(_context.t7);
            _context.next = 39;
            return _testUtils.default.sleepToEnsureDifferentTimestamps();

          case 39:
            _context.next = 41;
            return store.archiveDoc({
              id: created.id
            });

          case 41:
            archived = _context.sent;
            expect(archived).toEqual(Object.assign({}, updated, {
              archivedAt: archived.archivedAt,
              updatedAt: archived.updatedAt,
              order: _reorder.Reorder.DEFAULT_ORDER
            }));
            expect(archived.archivedAt).not.toBeFalsy();
            expect(archived.updatedAt).not.toEqual(updated.updatedAt); // Make sure timestamps aren't the same

            _context.next = 47;
            return _testUtils.default.sleepToEnsureDifferentTimestamps();

          case 47:
            _context.next = 49;
            return store.restoreDoc({
              id: created.id
            });

          case 49:
            restored = _context.sent;
            expect(restored).toEqual(Object.assign({}, updated, {
              updatedAt: restored.updatedAt,
              order: _reorder.Reorder.DEFAULT_ORDER
            }));
            expect(restored.archivedAt).toBeNull();
            expect(restored.updatedAt).not.toEqual(archived.updatedAt);

          case 53:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function shouldCRUD(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

exports.shouldCRUD = shouldCRUD;

var createDoc =
/*#__PURE__*/
function () {
  var _ref2 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee2(store, fieldValues) {
    var form;
    return _regenerator.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            form = createForm({
              value: fieldValues
            });
            return _context2.abrupt("return", store.createDoc({
              form: form,
              reorder: true
            }));

          case 2:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));

  return function createDoc(_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
}();

exports.createDoc = createDoc;

var createDocs =
/*#__PURE__*/
function () {
  var _ref3 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee3(store) {
    var harryValues, harry, hermioneValues, hermione, ronValues, ron;
    return _regenerator.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            harryValues = {
              firstName: 'Harry',
              lastName: 'Potter',
              order: 0
            };
            _context3.next = 3;
            return createDoc(store, harryValues);

          case 3:
            harry = _context3.sent;
            hermioneValues = {
              firstName: 'Hermione',
              lastName: 'Granger',
              order: 1
            };
            _context3.next = 7;
            return createDoc(store, hermioneValues);

          case 7:
            hermione = _context3.sent;
            _context3.next = 10;
            return store.archiveDoc({
              id: hermione.id
            });

          case 10:
            hermione = _context3.sent;
            ronValues = {
              firstName: 'Ron',
              lastName: 'Weasley'
            };
            _context3.next = 14;
            return createDoc(store, ronValues);

          case 14:
            ron = _context3.sent;
            return _context3.abrupt("return", {
              harry: harry,
              hermione: hermione,
              ron: ron
            });

          case 16:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));

  return function createDocs(_x5) {
    return _ref3.apply(this, arguments);
  };
}();

var updateDoc =
/*#__PURE__*/
function () {
  var _ref4 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee4(store, fieldValues) {
    var form;
    return _regenerator.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            form = createForm({
              value: fieldValues
            });
            return _context4.abrupt("return", store.updateDoc({
              form: form,
              reorder: true
            }));

          case 2:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));

  return function updateDoc(_x6, _x7) {
    return _ref4.apply(this, arguments);
  };
}();

exports.updateDoc = updateDoc;
var searchDefaults = {
  after: null,
  before: null,
  first: null,
  order: null,
  showArchived: null
};

var shouldGetAll =
/*#__PURE__*/
function () {
  var _ref5 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee5(Store, props) {
    var store, _ref6, harry, hermione, ron, all;

    return _regenerator.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            store = new Store(props);
            _context5.next = 3;
            return createDocs(store);

          case 3:
            _ref6 = _context5.sent;
            harry = _ref6.harry;
            hermione = _ref6.hermione;
            ron = _ref6.ron;
            all = {
              pageInfo: {
                hasNextPage: false
              },
              edges: [{
                node: harry
              }, {
                node: ron
              }, {
                node: hermione
              }]
            }; // Default props

            _context5.t0 = expect;
            _context5.next = 11;
            return store.getAllDocs((0, _objectSpread2.default)({}, searchDefaults, {
              showArchived: null
            }));

          case 11:
            _context5.t1 = _context5.sent;
            _context5.t2 = all;
            (0, _context5.t0)(_context5.t1).toEqual(_context5.t2);
            _context5.t3 = expect;
            _context5.next = 17;
            return store.getAllDocs((0, _objectSpread2.default)({}, searchDefaults, {
              where: {
                $and: [{
                  $or: [{
                    'fieldValues.firstName': {
                      $iLike: 'h%'
                    }
                  }, {
                    'fieldValues.lastName': {
                      $iLike: 'h%'
                    }
                  }]
                }]
              }
            }));

          case 17:
            _context5.t4 = _context5.sent;
            _context5.t5 = Object.assign({}, all, {
              edges: [{
                node: harry
              }, {
                node: hermione
              }]
            });
            (0, _context5.t3)(_context5.t4).toEqual(_context5.t5);
            _context5.t6 = expect;
            _context5.next = 23;
            return store.getAllDocs((0, _objectSpread2.default)({}, searchDefaults, {
              showArchived: true
            }));

          case 23:
            _context5.t7 = _context5.sent;
            _context5.t8 = Object.assign({}, all, {
              edges: [{
                node: hermione
              }]
            });
            (0, _context5.t6)(_context5.t7).toEqual(_context5.t8);
            _context5.t9 = expect;
            _context5.next = 29;
            return store.getAllDocs((0, _objectSpread2.default)({}, searchDefaults, {
              showArchived: false
            }));

          case 29:
            _context5.t10 = _context5.sent;
            _context5.t11 = Object.assign({}, all, {
              edges: [{
                node: harry
              }, {
                node: ron
              }]
            });
            (0, _context5.t9)(_context5.t10).toEqual(_context5.t11);
            _context5.t12 = expect;
            _context5.next = 35;
            return store.getAllDocs((0, _objectSpread2.default)({}, searchDefaults, {
              order: [['fieldValues.firstName', 'DESC']]
            }));

          case 35:
            _context5.t13 = _context5.sent;
            _context5.t14 = Object.assign({}, all, {
              edges: [{
                node: ron
              }, {
                node: hermione
              }, {
                node: harry
              }]
            });
            (0, _context5.t12)(_context5.t13).toEqual(_context5.t14);

          case 38:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));

  return function shouldGetAll(_x8, _x9) {
    return _ref5.apply(this, arguments);
  };
}();

exports.shouldGetAll = shouldGetAll;

var expectDocsToEqual = function expectDocsToEqual(docs, items) {
  var edges = [];
  items.forEach(function (item) {
    edges.push({
      node: {
        fieldValues: {
          firstName: item.firstName
        },
        order: item.order
      }
    });
  });
  expect(docs).toMatchObject({
    edges: edges
  });
};

var shouldMove =
/*#__PURE__*/
function () {
  var _ref7 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee6(Store, props) {
    var store, _ref8, harry, hermione, ron, all, ginnyValues, ginny;

    return _regenerator.default.wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            store = new Store(props);
            _context6.next = 3;
            return createDocs(store);

          case 3:
            _ref8 = _context6.sent;
            harry = _ref8.harry;
            hermione = _ref8.hermione;
            ron = _ref8.ron;
            all = {
              pageInfo: {
                hasNextPage: false
              },
              edges: [{
                node: harry
              }, {
                node: ron
              }, {
                node: hermione
              }]
            }; // Initial ordering

            _context6.t0 = expect;
            _context6.next = 11;
            return store.getAllDocs(searchDefaults);

          case 11:
            _context6.t1 = _context6.sent;
            _context6.t2 = all;
            (0, _context6.t0)(_context6.t1).toEqual(_context6.t2);
            // Create
            ginnyValues = {
              firstName: 'Ginny',
              lastName: 'Weasley',
              order: 1
            };
            _context6.next = 17;
            return createDoc(store, ginnyValues);

          case 17:
            ginny = _context6.sent;
            _context6.t3 = expectDocsToEqual;
            _context6.next = 21;
            return store.getAllDocs(searchDefaults);

          case 21:
            _context6.t4 = _context6.sent;
            _context6.t5 = [{
              firstName: 'Harry',
              order: 0
            }, {
              firstName: 'Ginny',
              order: 1
            }, {
              firstName: 'Ron',
              order: 2
            }, {
              firstName: 'Hermione',
              order: _reorder.Reorder.DEFAULT_ORDER
            }];
            (0, _context6.t3)(_context6.t4, _context6.t5);
            _context6.next = 26;
            return updateDoc(store, Object.assign({}, ron.fieldValues, {
              id: ron.id,
              order: 0
            }));

          case 26:
            ron = _context6.sent;
            _context6.t6 = expectDocsToEqual;
            _context6.next = 30;
            return store.getAllDocs(searchDefaults);

          case 30:
            _context6.t7 = _context6.sent;
            _context6.t8 = [{
              firstName: 'Ron',
              order: 0
            }, {
              firstName: 'Harry',
              order: 1
            }, {
              firstName: 'Ginny',
              order: 2
            }, {
              firstName: 'Hermione',
              order: _reorder.Reorder.DEFAULT_ORDER
            }];
            (0, _context6.t6)(_context6.t7, _context6.t8);
            _context6.next = 35;
            return updateDoc(store, Object.assign({}, harry.fieldValues, {
              id: harry.id,
              order: 2
            }));

          case 35:
            harry = _context6.sent;
            _context6.t9 = expectDocsToEqual;
            _context6.next = 39;
            return store.getAllDocs(searchDefaults);

          case 39:
            _context6.t10 = _context6.sent;
            _context6.t11 = [{
              firstName: 'Ron',
              order: 0
            }, {
              firstName: 'Ginny',
              order: 1
            }, {
              firstName: 'Harry',
              order: 2
            }, {
              firstName: 'Hermione',
              order: _reorder.Reorder.DEFAULT_ORDER
            }];
            (0, _context6.t9)(_context6.t10, _context6.t11);
            _context6.next = 44;
            return store.archiveDoc({
              id: ginny.id,
              reorder: true
            });

          case 44:
            _context6.t12 = expectDocsToEqual;
            _context6.next = 47;
            return store.getAllDocs(searchDefaults);

          case 47:
            _context6.t13 = _context6.sent;
            _context6.t14 = [{
              firstName: 'Ron',
              order: 0
            }, {
              firstName: 'Harry',
              order: 1
            }, {
              firstName: 'Hermione',
              order: _reorder.Reorder.DEFAULT_ORDER
            }, {
              firstName: 'Ginny',
              order: _reorder.Reorder.DEFAULT_ORDER
            }];
            (0, _context6.t12)(_context6.t13, _context6.t14);
            _context6.next = 52;
            return store.restoreDoc({
              id: ginny.id,
              reorder: true
            });

          case 52:
            _context6.t15 = expectDocsToEqual;
            _context6.next = 55;
            return store.getAllDocs(searchDefaults);

          case 55:
            _context6.t16 = _context6.sent;
            _context6.t17 = [{
              firstName: 'Ron',
              order: 0
            }, {
              firstName: 'Harry',
              order: 1
            }, {
              firstName: 'Ginny',
              order: 2
            }, {
              firstName: 'Hermione',
              order: _reorder.Reorder.DEFAULT_ORDER
            }];
            (0, _context6.t15)(_context6.t16, _context6.t17);

          case 58:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));

  return function shouldMove(_x10, _x11) {
    return _ref7.apply(this, arguments);
  };
}();

exports.shouldMove = shouldMove;
