// mson/lib/grid-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _component = _interopRequireDefault(require("./component"));

var GridItem =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(GridItem, _Component);

  function GridItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, GridItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(GridItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'GridItem');
    return _this;
  }

  (0, _createClass2.default)(GridItem, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(GridItem.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'hidden',
            component: 'BooleanField'
          }, {
            name: 'content',
            component: 'Field'
          }, {
            name: 'xl',
            component: 'IntegerField'
          }, {
            name: 'lg',
            component: 'IntegerField'
          }, {
            name: 'md',
            component: 'IntegerField'
          }, {
            name: 'sm',
            component: 'IntegerField'
          }, {
            name: 'xs',
            component: 'IntegerField'
          }]
        }
      });
    }
  }, {
    key: "set",
    value: function set(props) {
      if (props.content !== undefined) {
        props.content.set({
          parent: this
        });
      }

      (0, _get2.default)((0, _getPrototypeOf3.default)(GridItem.prototype), "set", this).call(this, props);
    }
  }]);
  return GridItem;
}(_component.default);

exports.default = GridItem;
