// mson/lib/infinite-loader.fixtures.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createMockedStore = exports.onGetItemElementMock = exports.onGetAllPeople = exports.allRecords = exports.records3 = exports.bowieFlat = exports.bowie = exports.michaelFlat = exports.michael = exports.records2 = exports.sinatraFlat = exports.sinatra = exports.stevieFlat = exports.stevie = exports.records1 = exports.ellaFlat = exports.ella = exports.rayFlat = exports.ray = exports.noop = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _testUtils = _interopRequireDefault(require("./test-utils"));

var _mapa = _interopRequireDefault(require("./mapa"));

var noop = function noop() {};

exports.noop = noop;

var defaults = _testUtils.default.toDefaultFieldsObject(null);

var ray = {
  node: {
    id: 'ray',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Ray'
    })
  },
  cursor: 'rayCursor'
};
exports.ray = ray;
var rayFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'ray',
  name: 'Ray'
});
exports.rayFlat = rayFlat;
var ella = {
  node: {
    id: 'ella',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Ella'
    })
  },
  cursor: 'ellaCursor'
};
exports.ella = ella;
var ellaFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'ella',
  name: 'Ella'
});
exports.ellaFlat = ellaFlat;
var records1 = {
  edges: [ray, ella]
};
exports.records1 = records1;
var stevie = {
  node: {
    id: 'stevie',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Stevie'
    })
  },
  cursor: 'stevieCursor'
};
exports.stevie = stevie;
var stevieFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'stevie',
  name: 'Stevie'
});
exports.stevieFlat = stevieFlat;
var sinatra = {
  node: {
    id: 'sinatra',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Sinatra'
    })
  },
  cursor: 'sinatraCursor'
};
exports.sinatra = sinatra;
var sinatraFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'sinatra',
  name: 'Sinatra'
});
exports.sinatraFlat = sinatraFlat;
var records2 = {
  edges: [stevie, sinatra]
};
exports.records2 = records2;
var michael = {
  node: {
    id: 'michael',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Michael'
    })
  },
  cursor: 'michaelCursor'
};
exports.michael = michael;
var michaelFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'michael',
  name: 'Michael'
});
exports.michaelFlat = michaelFlat;
var bowie = {
  node: {
    id: 'bowie',
    fieldValues: (0, _objectSpread2.default)({}, defaults, {
      name: 'Bowie'
    })
  },
  cursor: 'bowieCursor'
};
exports.bowie = bowie;
var bowieFlat = (0, _objectSpread2.default)({}, defaults, {
  id: 'bowie',
  name: 'Bowie'
});
exports.bowieFlat = bowieFlat;
var records3 = {
  edges: [michael, bowie]
};
exports.records3 = records3;
var noEdges = {
  pageInfo: {
    hasNextPage: false
  },
  edges: []
};
var allRecords = new _mapa.default();
exports.allRecords = allRecords;
allRecords.set(records1.edges[0].node.id, (0, _objectSpread2.default)({
  i: 0
}, records1.edges[0]));
allRecords.set(records1.edges[1].node.id, (0, _objectSpread2.default)({
  i: 1
}, records1.edges[1]));
allRecords.set(records2.edges[0].node.id, (0, _objectSpread2.default)({
  i: 2
}, records2.edges[0]));
allRecords.set(records2.edges[1].node.id, (0, _objectSpread2.default)({
  i: 3
}, records2.edges[1]));
allRecords.set(records3.edges[0].node.id, (0, _objectSpread2.default)({
  i: 4
}, records3.edges[0]));
allRecords.set(records3.edges[1].node.id, (0, _objectSpread2.default)({
  i: 5
}, records3.edges[1]));

var onGetAllPeople =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(props) {
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (props.before) {
              _context.next = 10;
              break;
            }

            _context.t0 = props.after;
            _context.next = _context.t0 === records1.edges[1].cursor ? 4 : _context.t0 === records2.edges[1].cursor ? 5 : _context.t0 === records3.edges[1].cursor ? 6 : 7;
            break;

          case 4:
            return _context.abrupt("return", records2);

          case 5:
            return _context.abrupt("return", records3);

          case 6:
            return _context.abrupt("return", noEdges);

          case 7:
            return _context.abrupt("return", records1);

          case 8:
            _context.next = 15;
            break;

          case 10:
            _context.t1 = props.before;
            _context.next = _context.t1 === records2.edges[0].cursor ? 13 : 14;
            break;

          case 13:
            return _context.abrupt("return", records1);

          case 14:
            return _context.abrupt("return", records2);

          case 15:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function onGetAllPeople(_x) {
    return _ref.apply(this, arguments);
  };
}();

exports.onGetAllPeople = onGetAllPeople;

var onGetItemElementMock = function onGetItemElementMock(id) {
  var i = allRecords.get(id).i;
  return {
    offsetTop: i * 100,
    offsetHeight: 100
  };
}; // const asyncNoop = async () => {};


exports.onGetItemElementMock = onGetItemElementMock;

var createMockedStore = function createMockedStore() {
  return {
    getAllDocs: function () {
      var _getAllDocs = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(props) {
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return onGetAllPeople(props);

              case 2:
                return _context2.abrupt("return", _context2.sent);

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      function getAllDocs(_x2) {
        return _getAllDocs.apply(this, arguments);
      }

      return getAllDocs;
    }(),
    // createDoc: asyncNoop,
    updateDoc: function () {
      var _updateDoc = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(props) {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                return _context3.abrupt("return", {
                  id: props.form.getValue('id')
                });

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3);
      }));

      function updateDoc(_x3) {
        return _updateDoc.apply(this, arguments);
      }

      return updateDoc;
    }(),
    // archiveDoc: asyncNoop,
    // restoreDoc: asyncNoop,
    on: noop
  };
};

exports.createMockedStore = createMockedStore;
