// mson/lib/component/mson-component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _wrappedComponent = _interopRequireDefault(require("./wrapped-component"));

var _registrar = _interopRequireDefault(require("../compiler/registrar"));

// Allows you to define to a component in MSON, which isn't built until the component is
// instantiated
var MSONComponent =
/*#__PURE__*/
function (_WrappedComponent) {
  (0, _inherits2.default)(MSONComponent, _WrappedComponent);

  function MSONComponent() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, MSONComponent);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(MSONComponent)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'MSONComponent');
    return _this;
  }

  (0, _createClass2.default)(MSONComponent, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(MSONComponent.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'definition',
            component: 'Field',
            required: true
          }]
        }
      }); // We want the class name to be that of the wrapped component

      this._preserveClassName = false; // For mocking

      this._registrar = _registrar.default;
    } // Set the compiler so that we can have separate compiler spaces per list of components

  }, {
    key: "_setCompiler",
    value: function _setCompiler(compiler) {
      this._compiler = compiler;
    }
  }, {
    key: "_getCompiler",
    value: function _getCompiler() {
      var compiler = this._compiler ? this._compiler : this._registrar.compiler;

      if (compiler) {
        return compiler;
      } else {
        throw new Error('compiler has not been registered with registrar');
      }
    }
  }, {
    key: "set",
    value: function set(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(MSONComponent.prototype), "set", this).call(this, props);

      if (props.definition !== undefined) {
        var compiler = this._getCompiler();

        var component = compiler.newComponent(props.definition);
        this.setComponentToWrap(component);
      }
    }
  }]);
  return MSONComponent;
}(_wrappedComponent.default);

exports.default = MSONComponent;
