// mson-react/lib/app-container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _app = _interopRequireDefault(require("./app"));

var _CssBaseline = _interopRequireDefault(require("@material-ui/core/CssBaseline"));

var _MuiThemeProvider = _interopRequireDefault(require("@material-ui/core/styles/MuiThemeProvider"));

var _createMuiTheme = _interopRequireDefault(require("@material-ui/core/styles/createMuiTheme"));

var _blueGrey = _interopRequireDefault(require("@material-ui/core/colors/blueGrey"));

var _lightBlue = _interopRequireDefault(require("@material-ui/core/colors/lightBlue"));

var _reactRouterDom = require("react-router-dom");

var _globals = _interopRequireDefault(require("mson/lib/globals"));

var _attach = _interopRequireDefault(require("./attach"));

var theme = (0, _createMuiTheme.default)({
  palette: {
    primary: _blueGrey.default,
    // primary: blue,
    // type: 'dark',
    // secondary: cyan
    secondary: _lightBlue.default
  },
  // TODO: remove after next major Material-UI release
  // (https://material-ui.com/style/typography/#migration-to-typography-v2)
  typography: {
    useNextVariants: true // shadows: ['none']

  }
}); // Note: BrowserRouter needs to be outside of App so that we can use withRouter

var AppContainer =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(AppContainer, _React$Component);

  function AppContainer() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, AppContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(AppContainer)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onNavigate", function (message, callback) {
      _globals.default.onNavigate(message, callback);
    });
    return _this;
  }

  (0, _createClass2.default)(AppContainer, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          component = _this$props.component,
          basename = _this$props.basename;
      return _react.default.createElement(_MuiThemeProvider.default, {
        theme: theme
      }, _react.default.createElement(_CssBaseline.default, null), _react.default.createElement(_reactRouterDom.BrowserRouter, {
        getUserConfirmation: this.onNavigate,
        basename: basename
      }, _react.default.createElement("div", null, _react.default.createElement(_app.default, {
        component: component
      }), _react.default.createElement(_reactRouterDom.Prompt, {
        message: "foo"
      }))));
    }
  }]);
  return AppContainer;
}(_react.default.Component);

var _default = (0, _attach.default)(['basename'])(AppContainer);

exports.default = _default;
