// mson-react/lib/button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _icon = _interopRequireDefault(require("./icon"));

var styles = function styles(theme) {
  return {
    leftIcon: {
      marginRight: theme.spacing.unit
    },
    marginTop: {
      marginTop: theme.spacing.unit * 3
    }
  };
};

var Button =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Button, _React$PureComponent);

  function Button() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Button);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Button)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function () {
      if (_this.props.onClick) {
        _this.props.onClick();
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Button, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          classes = _this$props.classes,
          type = _this$props.type,
          disabled = _this$props.disabled,
          label = _this$props.label,
          icon = _this$props.icon,
          fullWidth = _this$props.fullWidth,
          variant = _this$props.variant,
          marginTop = _this$props.marginTop;
      var className = marginTop !== false ? classes.marginTop : null;
      return _react.default.createElement(_Button.default, {
        className: className,
        type: type,
        color: "primary",
        disabled: disabled,
        onClick: this.handleClick,
        fullWidth: fullWidth,
        variant: variant
      }, icon ? _react.default.createElement(_icon.default, {
        className: classes.leftIcon,
        icon: icon
      }) : null, label);
    }
  }]);
  return Button;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(Button);

exports.default = _default;
