// mson-react/lib/card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _Grid = _interopRequireDefault(require("@material-ui/core/Grid"));

var _Paper = _interopRequireDefault(require("@material-ui/core/Paper"));

var _component = _interopRequireDefault(require("./component"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var styles = function styles(theme) {
  return {
    paper: {
      margin: theme.spacing.unit,
      padding: theme.spacing.unit * 2
    },
    content: {
      flex: 1 // TODO: needed?

    }
  };
};

var Card =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Card, _React$PureComponent);

  function Card() {
    (0, _classCallCheck2.default)(this, Card);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Card).apply(this, arguments));
  }

  (0, _createClass2.default)(Card, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          classes = _this$props.classes,
          component = _this$props.component;
      var content = component.get('content');
      var title = component.get('title');
      return _react.default.createElement(_Grid.default, {
        container: true,
        spacing: 0,
        justify: "center"
      }, _react.default.createElement(_Grid.default, {
        item: true,
        xs: 12,
        sm: 9,
        md: 6,
        lg: 6
      }, _react.default.createElement(_Paper.default, {
        className: classes.paper
      }, _react.default.createElement(_Grid.default, {
        container: true,
        wrap: "nowrap"
      }, _react.default.createElement(_Grid.default, {
        item: true,
        className: classes.content
      }, _react.default.createElement(_Typography.default, {
        variant: "h5",
        component: "h2"
      }, title), _react.default.createElement(_component.default, {
        component: content
      }))))));
    }
  }]);
  return Card;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(Card);

exports.default = _default;
