// mson-react/lib/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _innerComponent = _interopRequireDefault(require("./inner-component"));

var _compiler = _interopRequireDefault(require("mson/lib/compiler"));

var Component =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Component, _React$PureComponent);

  function Component() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Component);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Component)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      component: null
    });
    return _this;
  }

  (0, _createClass2.default)(Component, [{
    key: "createComponent",
    value: function createComponent() {
      var definition = this.props.definition;
      this.setState({
        component: _compiler.default.newComponent(definition)
      });
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.definition) {
        this.createComponent();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var component = this.state.component; // Did we use a definition to define a component in the state?

      if (component) {
        // Fire the unmount before we destroy the component so that the event is not lost
        component.emitChange('unmount'); // Remove all listeners to prevent listener leaks

        this.state.component.destroy();
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      // Is the definition changing?
      if (this.props.definition !== prevProps.definition) {
        this.createComponent();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          definition = _this$props.definition,
          component = _this$props.component,
          childProps = (0, _objectWithoutProperties2.default)(_this$props, ["definition", "component"]);
      var comp = component; // Should we use the component generated from the definition?

      if (!component) {
        comp = this.state.component;
      }

      return _react.default.createElement(_innerComponent.default, (0, _extends2.default)({
        component: comp
      }, childProps));
    }
  }]);
  return Component;
}(_react.default.PureComponent);

exports.default = Component;
