// mson-react/lib/field-editor-form.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _Card = _interopRequireDefault(require("@material-ui/core/Card"));

var _CardContent = _interopRequireDefault(require("@material-ui/core/CardContent"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _form = _interopRequireDefault(require("./form"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _field = _interopRequireDefault(require("./fields/field"));

var _compiler = _interopRequireDefault(require("mson/lib/compiler"));

var _each = _interopRequireDefault(require("lodash/each"));

var _attach = _interopRequireDefault(require("./attach"));

var _componentField = _interopRequireDefault(require("mson/lib/fields/component-field"));

var styles = function styles(theme) {
  return {
    form: {
      overflowY: 'scroll',
      maxHeight: 'calc(100vh - 390px)',
      marginTop: theme.spacing.unit * 3
    }
  };
};

var FieldEditorForm =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(FieldEditorForm, _React$PureComponent);

  function FieldEditorForm() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FieldEditorForm);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FieldEditorForm)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      field: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleValueChange", function (value) {
      var form = _this.props.component;
      var field = _this.state.field;
      var component = null; // Is the field changing?

      if (!field || value.componentName !== field.getClassName()) {
        if (field) {
          // Prevent listener leak
          field.destroy();
        }

        if (value.componentName) {
          component = _compiler.default.newComponent({
            component: value.componentName
          }); // Is the component a field?

          if (component.isField) {
            field = component;
          } else {
            // e.g. Text component is not a field
            field = new _componentField.default({
              content: component
            });
          } // Auto validate so that the user can preview how the validation will work


          var validate = function validate() {
            field.clearErr();
            field.validate();
          };

          field.on('value', validate);
          field.on('touched', validate);
        } else {
          field = null;
        }
      }

      if (field) {
        // Set field values using form values. The corresponding values may not exist if the field was
        // just changed.
        var values = form.getValues({
          default: false
        });
        component = field.getClassName() === 'ComponentField' ? field.get('content') : field;
        (0, _each.default)(values, function (value, name) {
          if (component.has(name)) {
            component.set((0, _defineProperty2.default)({}, name, value));
          }
        });
      }

      _this.setState({
        field: field
      });
    });
    return _this;
  }

  (0, _createClass2.default)(FieldEditorForm, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      // Note: we use a listener on value instead of attaching to value as a shallow compare of value
      // would not trigger a state change.
      this.props.component.on('value', this.handleValueChange); // Initialize with the current value

      this.handleValueChange(this.props.component.getValues());
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.props.component.removeListener('value', this.handleValueChange);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          component = _this$props.component,
          classes = _this$props.classes,
          formTag = _this$props.formTag,
          mode = _this$props.mode;
      var field = this.state.field;
      var preview = null;

      if (field && mode !== 'read') {
        preview = _react.default.createElement(_Card.default, {
          className: classes.card
        }, _react.default.createElement(_CardContent.default, null, _react.default.createElement(_Typography.default, {
          color: "textSecondary",
          gutterBottom: true
        }, "Preview"), _react.default.createElement(_field.default, {
          component: field
        })));
      }

      if (mode !== 'read') {
        return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
          className: classes.preview
        }, preview), _react.default.createElement("div", {
          className: classes.form
        }, _react.default.createElement(_form.default, {
          component: component,
          formTag: formTag,
          mode: mode
        })));
      } else if (field) {
        // Disable so that the user can edit the field with a click
        return _react.default.createElement(_field.default, {
          component: field,
          disabled: true
        });
      } else {
        return null;
      }
    }
  }]);
  return FieldEditorForm;
}(_react.default.PureComponent);

FieldEditorForm = (0, _withStyles.default)(styles)(FieldEditorForm);
FieldEditorForm = (0, _attach.default)(['mode'])(FieldEditorForm);
var _default = FieldEditorForm;
exports.default = _default;
