// mson-react/lib/flex-break.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = function styles(theme) {
  return {
    // break: {
    //   flexBasis: '100%',
    //   width: '0px',
    //   height: '0px',
    //   overflow: 'hidden'
    // }
    break: {
      width: '100%'
    }
  };
};

var FlexBreak =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(FlexBreak, _React$PureComponent);

  function FlexBreak() {
    (0, _classCallCheck2.default)(this, FlexBreak);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FlexBreak).apply(this, arguments));
  }

  (0, _createClass2.default)(FlexBreak, [{
    key: "render",
    value: function render() {
      var classes = this.props.classes;
      return _react.default.createElement("div", {
        className: classes.break
      });
    }
  }]);
  return FlexBreak;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(FlexBreak);

exports.default = _default;
