// mson-react/lib/form-dialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _button = _interopRequireDefault(require("./button"));

var _Dialog = _interopRequireDefault(require("@material-ui/core/Dialog"));

var _DialogActions = _interopRequireDefault(require("@material-ui/core/DialogActions"));

var _withMobileDialog = _interopRequireDefault(require("@material-ui/core/withMobileDialog"));

var _DialogContent = _interopRequireDefault(require("@material-ui/core/DialogContent"));

var _component = _interopRequireDefault(require("./component"));

var _attach = _interopRequireDefault(require("./attach"));

var FormDialog =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(FormDialog, _React$PureComponent);

  function FormDialog() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FormDialog);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FormDialog)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      saveClicked: false,
      previousMode: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClose", function (withCancelButton) {
      // Prevent the user from losing data when pressing esc or clicking outside dialog
      var _this$props = _this.props,
          mode = _this$props.mode,
          onClose = _this$props.onClose,
          onCancel = _this$props.onCancel,
          currentForm = _this$props.currentForm;

      if (withCancelButton || mode !== 'update') {
        if (mode === 'update') {
          if (onCancel) {
            onCancel(currentForm);
          }
        } else {
          if (onClose) {
            onClose(currentForm);
          }
        }
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleEdit", function () {
      if (_this.props.onEdit) {
        _this.props.onEdit(_this.props.currentForm);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSave", function (event) {
      // Stop the form from refreshing the page
      event.preventDefault(); // Should the submit action be disabled? For example, we may be editing in a nested dialog

      if (!_this.props.disableSubmit) {
        if (_this.props.onSave) {
          _this.props.onSave();
        } // Disable the save button so that the user sees that something is being processed


        _this.setState({
          saveClicked: true
        });
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDelete", function () {
      if (_this.props.onDelete) {
        _this.props.onDelete(_this.props.currentForm);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(FormDialog, [{
    key: "isOpen",
    value: function isOpen() {
      return !!this.props.mode;
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      // If the mode or err changes then allow the user to click save
      if (this.props.mode !== prevProps.mode || this.props.err !== prevProps.err) {
        this.setState({
          saveClicked: false
        });
      }

      if (this.props.mode !== prevProps.mode) {
        this.setState({
          previousMode: prevProps.mode
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          mode = _this$props2.mode,
          component = _this$props2.component,
          forbidUpdate = _this$props2.forbidUpdate,
          forbidDelete = _this$props2.forbidDelete,
          value = _this$props2.value,
          fullScreen = _this$props2.fullScreen;
      var _this$state = this.state,
          saveClicked = _this$state.saveClicked,
          previousMode = _this$state.previousMode;
      var disableSave = component.hasErrorForTouchedField() || !component.get('dirty') || saveClicked;
      var open = this.isOpen();
      var buttons = null; // Note: we analyze the previousMode so that the user isn't flashed with new buttons immediately
      // after they click save or close the dialog

      if (mode === 'update' || mode === 'create' || mode === null && (previousMode === 'update' || previousMode === 'create')) {
        buttons = _react.default.createElement("div", null, _react.default.createElement(_button.default, {
          type: "submit",
          label: "Save",
          icon: "Save",
          disabled: disableSave
        }), _react.default.createElement(_button.default, {
          label: "Cancel",
          icon: "Cancel",
          onClick: function onClick() {
            return _this2.handleClose(true);
          }
        }));
      } else if (!forbidUpdate || !forbidDelete) {
        buttons = _react.default.createElement("div", null, forbidUpdate ? '' : _react.default.createElement(_button.default, {
          label: "Edit",
          icon: "Edit",
          onClick: this.handleEdit
        }), forbidDelete ? '' : _react.default.createElement(_button.default, {
          label: value.archivedAt ? 'Restore' : 'Delete',
          icon: value.archivedAt ? 'Restore' : 'Delete',
          onClick: this.handleDelete
        }), _react.default.createElement(_button.default, {
          label: "Close",
          icon: "Cancel",
          onClick: function onClick() {
            return _this2.handleClose(true);
          }
        }));
      }

      return _react.default.createElement(_Dialog.default // fullScreen will be true on a small screen
      , {
        fullScreen: fullScreen,
        open: open,
        onClose: function onClose() {
          return _this2.handleClose(false);
        },
        "aria-labelledby": "form-dialog-title"
      }, _react.default.createElement("form", {
        onSubmit: this.handleSave
      }, _react.default.createElement(_DialogContent.default, null, _react.default.createElement(_component.default, {
        component: component,
        formTag: false,
        mode: mode
      })), buttons ? _react.default.createElement(_DialogActions.default, null, buttons) : ''));
    }
  }]);
  return FormDialog;
}(_react.default.PureComponent);

FormDialog = (0, _withMobileDialog.default)()(FormDialog);
FormDialog = (0, _attach.default)(['err', 'dirty', 'value', 'mode', 'disableSubmit'])(FormDialog);
var _default = FormDialog;
exports.default = _default;
