// mson-react/lib/grid-item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("./attach"));

var _Grid = _interopRequireDefault(require("@material-ui/core/Grid"));

var _component = _interopRequireDefault(require("./component"));

var GridItem =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(GridItem, _React$PureComponent);

  function GridItem() {
    (0, _classCallCheck2.default)(this, GridItem);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(GridItem).apply(this, arguments));
  }

  (0, _createClass2.default)(GridItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          hidden = _this$props.hidden,
          content = _this$props.content,
          xl = _this$props.xl,
          lg = _this$props.lg,
          md = _this$props.md,
          sm = _this$props.sm,
          xs = _this$props.xs;

      if (hidden) {
        return null;
      } else {
        return _react.default.createElement(_Grid.default, {
          item: true,
          xl: xl,
          lg: lg,
          md: md,
          sm: sm,
          xs: xs
        }, _react.default.createElement(_component.default, {
          component: content
        }));
      }
    }
  }]);
  return GridItem;
}(_react.default.PureComponent);

var _default = (0, _attach.default)(['hidden', 'content', 'xl', 'lg', 'md', 'sm', 'xs'])(GridItem);

exports.default = _default;
