// mson-react/lib/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _Drawer = _interopRequireDefault(require("@material-ui/core/Drawer"));

var _Hidden = _interopRequireDefault(require("@material-ui/core/Hidden"));

var _Divider = _interopRequireDefault(require("@material-ui/core/Divider"));

var _submenu = _interopRequireDefault(require("./submenu"));

var _attach = _interopRequireDefault(require("./attach"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _registrar = _interopRequireDefault(require("mson/lib/compiler/registrar"));

// TODO: when click on another section, close other sections? Requires moving menu state to app?
var drawerWidth = 240;

var styles = function styles(theme) {
  return {
    drawerHeader: (0, _objectSpread2.default)({}, theme.mixins.toolbar, {
      paddingLeft: theme.spacing.unit * 3,
      paddingTop: theme.spacing.unit
    }),
    drawerPaper: {
      width: 250
    },
    drawPaperResponsive: (0, _defineProperty2.default)({}, theme.breakpoints.up('md'), {
      width: drawerWidth,
      // position: 'relative',
      // height: '100%',
      // Extend to bottom of screen. TODO: does this work on mobile?
      // https://stackoverflow.com/a/48506883/2831606
      position: 'fixed',
      height: '100vh'
    })
  };
};

var Menu =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Menu, _React$PureComponent);

  function Menu() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Menu);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Menu)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDrawerToggle", function () {
      if (_this.props.onDrawerToggle) {
        _this.props.onDrawerToggle();
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleNavigate", function (menuItem) {
      if (_this.props.onNavigate) {
        _this.props.onNavigate(menuItem);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Menu, [{
    key: "items",
    value: function items() {
      var _this2 = this;

      var _this$props = this.props,
          component = _this$props.component,
          path = _this$props.path;
      var items = component.get('items');
      var submenus = [];
      items.forEach(function (item, index) {
        // Has access to item?
        if ((!item.roles || _registrar.default.client && _registrar.default.client.user.hasRole(item.roles)) && item.hidden !== true) {
          submenus.push(_react.default.createElement(_submenu.default, {
            item: item,
            key: index,
            onNavigate: _this2.handleNavigate,
            path: path,
            onDrawerToggle: _this2.handleDrawerToggle
          }));
        }
      });
      return submenus;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          classes = _this$props2.classes,
          theme = _this$props2.theme,
          mobileOpen = _this$props2.mobileOpen,
          responsive = _this$props2.responsive;
      var items = null; // if (!roles || (registrar.client && registrar.client.user.hasRole(roles))) {

      items = this.items(); // }

      var drawer = _react.default.createElement("div", null, _react.default.createElement("div", {
        className: classes.drawerHeader
      }, _react.default.createElement(_Typography.default, {
        variant: "h4"
      }, "Logo")), _react.default.createElement(_Divider.default, null), items);

      var temporaryDrawer = _react.default.createElement(_Drawer.default, {
        variant: "temporary",
        anchor: theme.direction === 'rtl' ? 'right' : 'left',
        open: mobileOpen,
        classes: {
          paper: classes.drawerPaper
        },
        onClose: this.handleDrawerToggle,
        ModalProps: {
          keepMounted: true // Better open performance on mobile.

        }
      }, drawer);

      var permanentDrawer = _react.default.createElement(_Drawer.default, {
        variant: "permanent",
        open: true,
        classes: {
          paper: classes.drawerPaper + (responsive ? " ".concat(classes.drawPaperResponsive) : '')
        }
      }, drawer);

      if (responsive) {
        return _react.default.createElement("div", null, _react.default.createElement(_Hidden.default, {
          mdUp: true
        }, temporaryDrawer), _react.default.createElement(_Hidden.default, {
          smDown: true,
          implementation: "css"
        }, permanentDrawer));
      } else {
        return temporaryDrawer;
      }
    }
  }]);
  return Menu;
}(_react.default.PureComponent);

Menu = (0, _withStyles.default)(styles, {
  withTheme: true
})(Menu);
Menu = (0, _attach.default)(['items'
/*, 'roles'*/
])(Menu);
var _default = Menu;
exports.default = _default;
