// mson-react/lib/search-bar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _TextField = _interopRequireDefault(require("@material-ui/core/TextField"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _icon = _interopRequireDefault(require("./icon"));

var _classnames = _interopRequireDefault(require("classnames"));

// TODO: see https://material-ui-next.com/style/icons/#icons and implement:
// 1. Mouseover to change background color
var styles = function styles(theme) {
  return {
    textField: {
      width: '100%'
    },
    textFieldInput: {
      borderRadius: 3,
      backgroundColor: theme.palette.primary[400],
      fontSize: theme.typography.subtitle1.fontSize,
      padding: '10px 34px 10px 34px',
      marginTop: '2px',
      width: 'calc(100%)',
      color: theme.palette.common.white
    },
    // Needed so that relative components don't take up any space
    iconContainer: {
      position: 'relative',
      width: '100%',
      // Set 100% so that we can position the close icon relative to this
      height: '0px'
    },
    searchIcon: {
      position: 'absolute',
      top: '-31px',
      left: '6px'
    },
    closeIcon: {
      position: 'absolute',
      top: '-43px',
      right: '-5px'
    },
    grow: {
      flexGrow: 1
    }
  };
};

var SearchBar =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(SearchBar, _React$PureComponent);

  function SearchBar() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, SearchBar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(SearchBar)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      autoFullWidth: false
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyUp", function (event) {
      // Enter pressed?
      if (event.keyCode === 13) {
        _this.props.onSearch(_this.props.searchString);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClearSearch", function (event) {
      _this.props.onSearch('');
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleFocus", function () {
      _this.setState({
        autoFullWidth: true
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function () {
      _this.setState({
        autoFullWidth: false
      });
    });
    return _this;
  }

  (0, _createClass2.default)(SearchBar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          classes = _this$props.classes,
          className = _this$props.className,
          searchString = _this$props.searchString,
          _onChange = _this$props.onChange,
          fullWidth = _this$props.fullWidth;
      var autoFullWidth = this.state.autoFullWidth;
      var full = fullWidth || autoFullWidth;
      return _react.default.createElement("div", {
        className: (0, _classnames.default)(className, full ? classes.grow : null)
      }, _react.default.createElement(_TextField.default, {
        InputProps: {
          disableUnderline: true,
          classes: {
            input: classes.textFieldInput
          }
        },
        value: searchString,
        onKeyUp: this.handleKeyUp,
        onChange: function onChange(event) {
          return _onChange(event.target.value);
        },
        className: classes.textField,
        onFocus: this.handleFocus,
        onBlur: this.handleBlur
      }), _react.default.createElement("div", {
        className: classes.iconContainer
      }, _react.default.createElement(_icon.default, {
        icon: "Search",
        className: classes.searchIcon
      }), _react.default.createElement(_IconButton.default, {
        color: "inherit",
        "aria-label": "clear search",
        onClick: this.handleClearSearch,
        className: classes.closeIcon
      }, _react.default.createElement(_icon.default, {
        icon: "Close"
      }))));
    }
  }]);
  return SearchBar;
}(_react.default.PureComponent);

SearchBar.propTypes = {
  classes: _propTypes.default.object.isRequired
};
SearchBar = (0, _withStyles.default)(styles)(SearchBar);
var _default = SearchBar;
exports.default = _default;
