// mson-react/lib/snackbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _Snackbar = _interopRequireDefault(require("@material-ui/core/Snackbar"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _icon = _interopRequireDefault(require("./icon"));

// import Button from '@material-ui/core/Button';
var styles = function styles(theme) {
  return {
    close: {
      padding: theme.spacing.unit / 2
    }
  };
};

var Snackbar =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Snackbar, _React$PureComponent);

  function Snackbar() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Snackbar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Snackbar)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClose", function (event, reason) {
      if (reason === 'clickaway') {
        return;
      }

      if (_this.props.onClose) {
        _this.props.onClose();
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Snackbar, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          classes = _this$props.classes,
          message = _this$props.message,
          open = _this$props.open;
      return _react.default.createElement(_Snackbar.default, {
        anchorOrigin: {
          vertical: 'bottom',
          horizontal: 'left'
        },
        open: open,
        autoHideDuration: 6000,
        onClose: this.handleClose,
        ContentProps: {
          'aria-describedby': 'message-id'
        },
        message: _react.default.createElement("span", {
          id: "message-id"
        }, message),
        action: [// <Button key="undo" color="secondary" size="small" onClick={this.handleClose}>
        //   UNDO
        // </Button>,
        _react.default.createElement(_IconButton.default, {
          key: "close",
          "aria-label": "Close",
          color: "inherit",
          className: classes.close,
          onClick: this.handleClose
        }, _react.default.createElement(_icon.default, {
          icon: "Close"
        }))]
      });
    }
  }]);
  return Snackbar;
}(_react.default.PureComponent);

Snackbar.propTypes = {
  classes: _propTypes.default.object.isRequired
};

var _default = (0, _withStyles.default)(styles)(Snackbar);

exports.default = _default;
