// mson-react/lib/submenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var _List = _interopRequireDefault(require("@material-ui/core/List"));

var _ListItem = _interopRequireDefault(require("@material-ui/core/ListItem"));

var _ListItemText = _interopRequireDefault(require("@material-ui/core/ListItemText"));

var _Collapse = _interopRequireDefault(require("@material-ui/core/Collapse"));

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _icon = _interopRequireDefault(require("./icon"));

var _each = _interopRequireDefault(require("lodash/each"));

var _registrar = _interopRequireDefault(require("mson/lib/compiler/registrar"));

var styles = function styles(theme) {
  return {
    nested: {
      paddingLeft: theme.spacing.unit * 4
    },
    primary: {
      color: theme.palette.text.primary
    },
    secondary: {
      color: theme.palette.text.secondary
    },
    selected: {
      fontWeight: 'bold'
    }
  };
};

var Submenu =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(Submenu, _React$PureComponent);

  function Submenu(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Submenu);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Submenu).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (item) {
      // Has sub items?
      if (item.items) {
        _this.setState({
          open: !_this.state.open
        });
      } else {
        // Note: the navigation is handled by the parent as we want to be able to prevent the
        // navigation if say a form is dirty, etc...
        if (_this.props.onNavigate) {
          _this.props.onNavigate(item);
        }

        if (_this.props.onDrawerToggle) {
          // Close the drawer
          _this.props.onDrawerToggle();
        }
      }
    });
    _this.state = {
      open: _this.childSelected()
    };
    return _this;
  }

  (0, _createClass2.default)(Submenu, [{
    key: "childSelected",
    value: function childSelected() {
      var _this$props = this.props,
          item = _this$props.item,
          path = _this$props.path;
      var childSelected = false;
      (0, _each.default)(item.items, function (item) {
        if (path === item.path) {
          childSelected = true;
          return false; // exit loop
        }
      });
      return childSelected;
    }
  }, {
    key: "items",
    value: function items() {
      var _this2 = this;

      var _this$props2 = this.props,
          classes = _this$props2.classes,
          item = _this$props2.item,
          path = _this$props2.path;
      var listItems = [];
      item.items.forEach(function (item, index) {
        var isSelected = path === item.path;
        var classNames = [classes.secondary];

        if (isSelected) {
          classNames.push(classes.selected);
        }

        if ((!item.roles || _registrar.default.client && _registrar.default.client.user.hasRole(item.roles)) && item.hidden !== true) {
          listItems.push(_react.default.createElement(_ListItem.default, {
            button: true,
            className: classes.nested,
            key: index,
            onClick: function onClick() {
              return _this2.handleClick(item);
            }
          }, _react.default.createElement(_ListItemText.default, {
            disableTypography: true,
            primary: _react.default.createElement(_Typography.default, {
              variant: "body2",
              className: classNames.join(' ')
            }, item.label)
          })));
        }
      });
      return listItems;
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props3 = this.props,
          item = _this$props3.item,
          classes = _this$props3.classes,
          path = _this$props3.path;
      var items = item.items ? this.items() : null;
      var isSelected = path === item.path;
      var classNames = [classes.primary];

      if (isSelected) {
        classNames.push(classes.selected);
      }

      var listItems = null;

      if (items) {
        listItems = _react.default.createElement(_Collapse.default, {
          in: this.state.open,
          timeout: "auto",
          unmountOnExit: true
        }, _react.default.createElement(_List.default, {
          disablePadding: true
        }, this.items()));
      }

      return _react.default.createElement("div", null, _react.default.createElement(_ListItem.default, {
        button: true,
        onClick: function onClick() {
          return _this3.handleClick(item);
        }
      }, _react.default.createElement(_ListItemText.default, {
        disableTypography: true,
        primary: _react.default.createElement(_Typography.default, {
          variant: "subtitle1",
          className: classNames.join(' ')
        }, item.label)
      }), items ? this.state.open ? _react.default.createElement(_icon.default, {
        icon: "ExpandLess"
      }) : _react.default.createElement(_icon.default, {
        icon: "ExpandMore"
      }) : null), listItems);
    }
  }]);
  return Submenu;
}(_react.default.PureComponent);

var _default = (0, _withStyles.default)(styles)(Submenu);

exports.default = _default;
