// mson-react/lib/tabs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _attach = _interopRequireDefault(require("./attach"));

var _AppBar = _interopRequireDefault(require("@material-ui/core/AppBar"));

var _Tabs = _interopRequireDefault(require("@material-ui/core/Tabs"));

var _Tab = _interopRequireDefault(require("@material-ui/core/Tab"));

var _icon = _interopRequireDefault(require("./icon"));

var Tabs =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(Tabs, _React$Component);

  function Tabs() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Tabs);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Tabs)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (event, value) {
      var _this$props = _this.props,
          component = _this$props.component,
          items = _this$props.items;
      component.set({
        value: value
      });
      var itemName = items[value].name;
      component.emitChange(itemName);
    });
    return _this;
  }

  (0, _createClass2.default)(Tabs, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          items = _this$props2.items,
          value = _this$props2.value; // FUTURE: option to make this fixed under the main app bar?

      return _react.default.createElement(_AppBar.default, {
        position: "static",
        color: "default",
        elevation: 1 // tone down the elevation>

      }, _react.default.createElement(_Tabs.default, {
        value: value,
        onChange: this.handleChange,
        indicatorColor: "primary",
        textColor: "primary",
        scrollButtons: "auto",
        variant: "fullWidth"
      }, items.map(function (item, index) {
        return _react.default.createElement(_Tab.default, {
          label: item.label,
          icon: _react.default.createElement(_icon.default, {
            icon: item.icon
          }),
          key: index
        });
      })));
    }
  }]);
  return Tabs;
}(_react.default.Component);

Tabs = (0, _attach.default)(['items', 'value'])(Tabs);
var _default = Tabs;
exports.default = _default;
