// mson-react/lib/user-menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _IconButton = _interopRequireDefault(require("@material-ui/core/IconButton"));

var _icon = _interopRequireDefault(require("./icon"));

var _MenuItem = _interopRequireDefault(require("@material-ui/core/MenuItem"));

var _Menu = _interopRequireDefault(require("@material-ui/core/Menu"));

var UserDropDown =
/*#__PURE__*/
function (_React$PureComponent) {
  (0, _inherits2.default)(UserDropDown, _React$PureComponent);

  function UserDropDown() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, UserDropDown);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(UserDropDown)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      anchorEl: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMenu", function (event) {
      _this.setState({
        anchorEl: event.currentTarget
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClose", function () {
      _this.setState({
        anchorEl: null
      });
    });
    return _this;
  }

  (0, _createClass2.default)(UserDropDown, [{
    key: "render",
    value: function render() {
      var isLoggedIn = this.props.isLoggedIn;
      var anchorEl = this.state.anchorEl;
      var open = Boolean(anchorEl);

      if (!isLoggedIn) {
        return null;
      }

      return _react.default.createElement("div", null, _react.default.createElement(_IconButton.default, {
        "aria-owns": open ? 'menu-appbar' : null,
        "aria-haspopup": "true",
        onClick: this.handleMenu,
        color: "inherit"
      }, _react.default.createElement(_icon.default, {
        icon: "AccountCircle"
      })), _react.default.createElement(_Menu.default, {
        id: "menu-appbar",
        anchorEl: anchorEl,
        anchorOrigin: {
          vertical: 'top',
          horizontal: 'right'
        },
        transformOrigin: {
          vertical: 'top',
          horizontal: 'right'
        },
        open: open,
        onClose: this.handleClose
      }, _react.default.createElement(_MenuItem.default, {
        onClick: this.handleClose
      }, "Log Out"), _react.default.createElement(_MenuItem.default, {
        onClick: this.handleClose
      }, "Edit Account"), _react.default.createElement(_MenuItem.default, {
        onClick: this.handleClose
      }, "Change Password")));
    }
  }]);
  return UserDropDown;
}(_react.default.PureComponent);

exports.default = UserDropDown;
