// mson/lib/form/field-editor-form.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _form = _interopRequireDefault(require("./form"));

var _selectField = _interopRequireDefault(require("../fields/select-field"));

var _each = _interopRequireDefault(require("lodash/each"));

var FieldEditorForm =
/*#__PURE__*/
function (_Form) {
  (0, _inherits2.default)(FieldEditorForm, _Form);

  function FieldEditorForm() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FieldEditorForm);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FieldEditorForm)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'FieldEditorForm');
    return _this;
  }

  (0, _createClass2.default)(FieldEditorForm, [{
    key: "_namesToOptions",
    value: function _namesToOptions() {
      return this.constructor.componentNames.map(function (name) {
        return {
          value: name,
          label: name
        };
      });
    }
  }, {
    key: "_clearFields",
    value: function _clearFields() {
      var _this2 = this;

      this.eachField(function (field) {
        var name = field.get('name');

        if (!_this2.isDefaultField(name) && name !== 'componentName') {
          _this2.removeField(name);
        }
      });

      this._emitChangeToFields(null);
    }
  }, {
    key: "_copySchemaFields",
    value: function _copySchemaFields(schema) {
      var _this3 = this;

      var fields = [];
      schema.eachField(function (field) {
        var name = field.get('name');

        if (!schema.isDefaultField(name) && field.get('docLevel') === 'basic') {
          _this3.addField(field);

          fields.push(field);
        }
      });

      this._emitChangeToFields(fields);
    }
  }, {
    key: "_adjustFields",
    value: function _adjustFields(values) {
      var _this4 = this;

      // Use the component schema to display fields to the user
      // Clear any previous fields that are not default fields or componentName
      this._clearFields();

      if (values.componentName) {
        var compiler = this._registrar.compiler;
        this._component = compiler.newComponent({
          component: values.componentName
        });
        var schema = new _form.default();

        this._component.buildSchemaForm(schema, compiler);

        this._copySchemaFields(schema); // Preserve any applicable values from the previous component


        (0, _each.default)(values, function (value, name) {
          if (_this4.hasField(name)) {
            _this4.getField(name).setValue(value);
          }
        });
      }
    }
  }, {
    key: "_create",
    value: function _create(props) {
      var _this5 = this;

      (0, _get2.default)((0, _getPrototypeOf3.default)(FieldEditorForm.prototype), "_create", this).call(this, props);
      this.set({
        fields: [new _selectField.default({
          name: 'componentName',
          label: 'Field',
          options: this._namesToOptions()
        })]
      }); // Adjust the fields when the componentName changes

      this.getField('componentName').on('value', function () {
        return _this5._adjustFields(_this5.getValues());
      });
    }
  }, {
    key: "setValues",
    value: function setValues(values) {
      // Is the componentName changing?
      if (!this._component || values.componentName && this._component.getClassName() !== values.componentName) {
        this._adjustFields(Object.assign({}, this.getValues(), {
          componentName: values.componentName
        }));
      }

      (0, _get2.default)((0, _getPrototypeOf3.default)(FieldEditorForm.prototype), "setValues", this).call(this, values);
    }
  }]);
  return FieldEditorForm;
}(_form.default);

exports.default = FieldEditorForm;
(0, _defineProperty2.default)(FieldEditorForm, "componentNames", ['AddressField', 'BooleanField', 'ButtonField', 'ChainedSelectField', 'ChainedSelectListField', 'CityField', // 'CollectionField',
// 'CompositeField',
'CountryField', 'DateField', 'EmailField', // 'FormField',
'IdField', 'IntegerField', // 'ListField',
'MoneyField', 'NumberField', 'PasswordField', 'PersonFullNameField', 'PersonNameField', 'PhoneField', 'ProvinceField', 'ReCAPTCHAField', // 'RegExpField',
'SelectField', 'SelectListField', 'StateField', 'Text', 'TextField', 'TextListField', 'TimeField', 'URLField']);
