// mson/lib/form/form-builder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _form = _interopRequireDefault(require("./form"));

var _formEditor = _interopRequireDefault(require("./form-editor"));

var _tabs = _interopRequireDefault(require("../tabs"));

var _formField = _interopRequireDefault(require("../fields/form-field"));

var _set = _interopRequireDefault(require("../actions/set"));

var FormBuilder =
/*#__PURE__*/
function (_Form) {
  (0, _inherits2.default)(FormBuilder, _Form);

  function FormBuilder() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FormBuilder);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FormBuilder)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'FormBuilder');
    return _this;
  }

  (0, _createClass2.default)(FormBuilder, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(FormBuilder.prototype), "_create", this).call(this, props);
      this.set({
        fields: [new _tabs.default({
          name: 'tabs',
          items: [{
            name: 'edit',
            label: 'Edit',
            icon: 'Edit'
          }, {
            name: 'preview',
            label: 'Preview',
            icon: 'ViewCompact'
          }]
        }), new _formField.default({
          name: 'form',
          form: new _formEditor.default()
        })],
        listeners: [{
          event: 'fields.tabs.content.edit',
          actions: [new _set.default({
            name: 'fields.form.form.editable',
            value: true
          })]
        }, {
          event: 'fields.tabs.content.preview',
          actions: [new _set.default({
            name: 'fields.form.form.editable',
            value: false
          })]
        }]
      });
    }
  }]);
  return FormBuilder;
}(_form.default);

exports.default = FormBuilder;
