// mson/lib/form/form-editor.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _form = _interopRequireDefault(require("./form"));

var _collectionField = _interopRequireDefault(require("../fields/collection-field"));

var _factory = _interopRequireDefault(require("../component/factory"));

var _fieldEditorForm = _interopRequireDefault(require("./field-editor-form"));

var FormEditor =
/*#__PURE__*/
function (_Form) {
  (0, _inherits2.default)(FormEditor, _Form);

  function FormEditor() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, FormEditor);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(FormEditor)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'FormEditor');
    return _this;
  }

  (0, _createClass2.default)(FormEditor, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(FormEditor.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'definition',
            component: 'Field'
          }]
        },
        fields: [new _collectionField.default({
          name: 'fields',
          label: 'Fields',
          hideLabel: true,
          maxColumns: 1,
          skipRead: true,
          includeExtraneous: true,
          forbidOrder: false,
          formFactory: new _factory.default({
            product: function product() {
              return new _fieldEditorForm.default();
            }
          })
        })]
      });
    }
  }, {
    key: "_setDefinition",
    value: function _setDefinition(definition) {
      var fields = definition.fields.map(function (field) {
        return (0, _objectSpread2.default)({}, field, {
          componentName: field.component
        });
      });
      this.get('fields.fields').setValue(fields);
    }
  }, {
    key: "set",
    value: function set(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(FormEditor.prototype), "set", this).call(this, (0, _objectSpread2.default)({}, props, {
        definition: undefined
      }));

      if (props.definition !== undefined) {
        this._setDefinition(props.definition);
      }
    }
  }, {
    key: "_getDefinition",
    value: function _getDefinition() {
      return {
        component: 'Form',
        fields: this.get('fields.fields').mapForms(function (form) {
          return (0, _objectSpread2.default)({}, form.getValues({
            default: false
          }), {
            component: form.getValue('componentName'),
            componentName: undefined
          });
        })
      };
    }
  }, {
    key: "getOne",
    value: function getOne(name) {
      if (name === 'definition') {
        return this._getDefinition();
      }

      return (0, _get2.default)((0, _getPrototypeOf3.default)(FormEditor.prototype), "getOne", this).call(this, name);
    }
  }]);
  return FormEditor;
}(_form.default);

exports.default = FormEditor;
