// mson/lib/form/schema-validator-form.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _form = _interopRequireDefault(require("./form"));

var SchemaValidatorForm =
/*#__PURE__*/
function (_Form) {
  (0, _inherits2.default)(SchemaValidatorForm, _Form);

  function SchemaValidatorForm() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, SchemaValidatorForm);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(SchemaValidatorForm)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'SchemaValidatorForm');
    return _this;
  }

  (0, _createClass2.default)(SchemaValidatorForm, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(SchemaValidatorForm.prototype), "_create", this).call(this, props);
      this.set({
        schema: {
          component: 'Form',
          fields: [{
            name: 'compiler',
            component: 'Field',
            required: true
          }]
        }
      });
    }
  }, {
    key: "setValues",
    value: function setValues(values) {
      // Dynamically build this form based on the component
      var compiler = this.get('compiler');
      var field = compiler.newComponent({
        component: values.component
      });
      field.buildSchemaForm(this, compiler); // We assign the name to the id so that errors are more descriptive

      (0, _get2.default)((0, _getPrototypeOf3.default)(SchemaValidatorForm.prototype), "setValues", this).call(this, Object.assign({}, values, {
        id: values.name
      }));
    }
  }]);
  return SchemaValidatorForm;
}(_form.default);

exports.default = SchemaValidatorForm;
