// mson/lib/form/user.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _form = _interopRequireDefault(require("./form"));

var _msonComponent = _interopRequireDefault(require("../component/mson-component"));

var User =
/*#__PURE__*/
function (_Form) {
  (0, _inherits2.default)(User, _Form);

  function User() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, User);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(User)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_className", 'User');
    return _this;
  }

  (0, _createClass2.default)(User, [{
    key: "_create",
    value: function _create(props) {
      (0, _get2.default)((0, _getPrototypeOf3.default)(User.prototype), "_create", this).call(this, props); // Note: we use MSONComponents as EmailField and PasswordField are uncompiled and importing from
      // ../fields would create a circular dependency.

      var fields = [new _msonComponent.default({
        definition: {
          component: 'EmailField',
          name: 'username',
          label: 'Email',
          required: true // in: false,

        }
      }), new _msonComponent.default({
        definition: {
          component: 'PasswordField',
          name: 'password',
          label: 'Password',
          // required: true, // Required by listeners when creating
          in: false,
          out: false,
          hidden: true,
          forbidSort: true
        }
      })];
      var schema = {
        component: 'Form',
        fields: [{
          name: 'roles',
          component: 'RolesField'
        }]
      }; // By default, lock down access

      var access = {
        form: {
          create: 'admin',
          read: 'admin',
          update: 'admin',
          archive: 'admin'
        }
      };
      this.set({
        schema: schema,
        access: access,
        fields: fields
      });
    }
  }]);
  return User;
}(_form.default);

exports.default = User;
